{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BCL2FASQ COMMAND"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#separates differenlty indexed samples. multiplexes i7. NOT dempultiplexing the wells but the lanes on 10x\n",
    "#there will be two sublibraries of the i7 indeces. Nextset\n",
    "#the scifi5 o_DT are the main seq containing cdna, WP are the AAVs\n",
    "#bcl2fastq -i ./Data/Intensities/BaseCalls/ -R ./ -o ./demultiplexed/ --use-bases-mask Y26,I8,Y17,Y115 --sample-sheet ../20220114_SampleSheet_HJCNYAFX3.csv -r 8 -p 24 -w 8 --no-lane-splitting --minimum-trimmed-read-length 10 --mask-short-adapter-reads 10 --ignore-missing-bcl --create-fastq-for-index-reads --ignore-missing-filter --ignore-missing-position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "from subprocess import call\n",
    "import subprocess\n",
    "import glob\n",
    "import pandas as pd\n",
    "import tempfile\n",
    "import ntpath\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#in_path = \"'/media/Dropbox/MNM network files/OrganizedSeqFiles/snRNA-aSyn/MNM220114/220114_NB502004_0177_AHJCNYAFX3/demultiplexed/SciFi_10X_Pilot5_Validation_R2'\"\n",
    "#out_path = \"'/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/output_10X_pilot5_r2'\"\n",
    "#base_path = '/tmp/indata'\n",
    "#export_path = '/tmp/outdata'\n",
    "in_path = \"'/media/data/AtteR/210517_A00681_0370_AHF2JHDRXY/demultiplexed'\"\n",
    "out_path = \"'/media/data/AtteR/scifi-analysis/pilot1_output'\"\n",
    "\n",
    "in_path = \"/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled\"\n",
    "out_path = \"/media/data/AtteR/scifi-analysis/scifi6/output\"\n",
    "base_path = '/tmp/indata'\n",
    "export_path = '/tmp/outdata'\n",
    "output_data2 = '/media/data/AtteR/scifi-analysis/scifi6' \n",
    "output_tables = '/media/data/AtteR/scifi-analysis/summary_tables'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/media/data/AtteR/scifi-analysis/scifi6/output’: File exists\n",
      "ln: failed to create symbolic link '/tmp/indata/SciFi_10X_Pilot6_Pooled': File exists\n"
     ]
    }
   ],
   "source": [
    "#the symlinks are generated to ensure path without spaces which bbmap cannot handle\n",
    "!mkdir $out_path\n",
    "#!rm $base_path\n",
    "!ln -s $in_path $base_path\n",
    "!rm $export_path\n",
    "!ln -s $out_path $export_path\n",
    "base_path = base_path +'/'\n",
    "export_path = export_path + '/'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fuse Scifi and Bead barcode reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fuseBCs(sample_id) :\n",
    "    print(sample_id) \n",
    "    revCompR1 = \"/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/\"+ os.path.basename(sample_id).replace('_R1_001.fastq.gz','_RC_R1_001.fastq.gz')\n",
    "\n",
    "    #revCompR1 = '/media/data/AtteR/scifi-analysis/'+ os.path.basename(sample_id).replace('_R1_001.fastq.gz','_RC_R1_001.fastq.gz')\n",
    "    rc_call= \"reformat.sh overwrite=t rcomp=t in='\"+ sample_id +\"' out='\"+revCompR1+\"'\"\n",
    "    print(rc_call)\n",
    "    !$rc_call\n",
    "    #reverse complement of the read1 needed prior to fusing it with with read2\n",
    "    inPathR2 = base_path + ntpath.basename(sample_id).replace('_R1_001.fastq.gz','_R2_001.fastq.gz')\n",
    "    outPath = base_path + ntpath.basename(sample_id).replace('_R1_001.fastq.gz','_R21_001.fastq.gz')\n",
    "    fuse_call= \"/usr/local/bin/bbmap/fuse.sh in1='\"+ inPathR2 +\"' in2='\"+revCompR1+\"' out='\"+outPath+\"' overwrite=t interleaved=f fusepairs=t pad=0\"\n",
    "    !$fuse_call  #BBMap is a splice-aware global aligner\n",
    "    os.remove(revCompR1)\n",
    "    print(fuse_call)\n",
    "    print(\"outPath: \" + outPath)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz', '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_R1_001.fastq.gz', '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_RC_R1_001.fastq.gz', '/tmp/indata/SciFi_new_cDNA_Pool1_R1_001.fastq.gz', '/tmp/indata/SciFi_old_cDNA_Pool2_R1_001.fastq.gz']\n",
      "SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz\n",
      "SciFi_Pool1-2_EnrichedBCs_R1_001.fastq.gz\n",
      "SciFi_Pool1-2_EnrichedBCs_RC_R1_001.fastq.gz\n",
      "SciFi_new_cDNA_Pool1_R1_001.fastq.gz\n",
      "SciFi_old_cDNA_Pool2_R1_001.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R1*.fastq.gz'))\n",
    "samples_list= [item for item in samples_list if 'Undetermined' not in item] \n",
    "#samples_list= [item for item in samples_list if 'EnrichedBCs' not in item] \n",
    "\n",
    "#samples_list= [item for item in samples_list if 'BC' in item]\n",
    "print(samples_list)\n",
    "for sample_id in samples_list:\n",
    "    print(ntpath.basename(sample_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz\n",
      "reformat.sh overwrite=t rcomp=t in='/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz' out='/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz'\n",
      "java -ea -Xms300m -cp /usr/local/bin/bbmap/current/ jgi.ReformatReads overwrite=t rcomp=t in=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz out=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz\n",
      "Executing jgi.ReformatReads [overwrite=t, rcomp=t, in=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz, out=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz]\n",
      "\n",
      "Set RCOMP to true\n",
      "Input is being processed as unpaired\n",
      "Input:                  \t84309009 reads          \t2192034234 bases\n",
      "Output:                 \t84309009 reads (100.00%) \t2192034234 bases (100.00%)\n",
      "\n",
      "Time:                         \t182.581 seconds.\n",
      "Reads Processed:      84309k \t461.76k reads/sec\n",
      "Bases Processed:       2192m \t12.01m bases/sec\n",
      "java -ea -Xmx195646m -cp /usr/local/bin/bbmap/current/ jgi.FuseSequence in1=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R2_001.fastq.gz in2=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz out=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz overwrite=t interleaved=f fusepairs=t pad=0\n",
      "Executing jgi.FuseSequence [in1=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R2_001.fastq.gz, in2=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz, out=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz, overwrite=t, interleaved=f, fusepairs=t, pad=0]\n",
      "\n",
      "Set INTERLEAVED to false\n",
      "Input is being processed as paired\n",
      "Writing interleaved.\n",
      "Time:                         \t48.488 seconds.\n",
      "Reads Processed:        168m \t3477.54k reads/sec\n",
      "Bases Processed:       3625m \t74.77m bases/sec\n",
      "/usr/local/bin/bbmap/fuse.sh in1='/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R2_001.fastq.gz' in2='/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz' out='/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz' overwrite=t interleaved=f fusepairs=t pad=0\n",
      "outPath: /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz\n",
      "/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz\n",
      "reformat.sh overwrite=t rcomp=t in='/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz' out='/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz'\n",
      "java -ea -Xms300m -cp /usr/local/bin/bbmap/current/ jgi.ReformatReads overwrite=t rcomp=t in=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz out=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz\n",
      "Executing jgi.ReformatReads [overwrite=t, rcomp=t, in=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz, out=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz]\n",
      "\n",
      "Set RCOMP to true\n",
      "Exception in thread \"main\" java.lang.RuntimeException: Can't read file '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R1_001.fastq.gz'\n",
      "\tat shared.Tools.testInputFiles(Tools.java:185)\n",
      "\tat jgi.ReformatReads.<init>(ReformatReads.java:358)\n",
      "\tat jgi.ReformatReads.main(ReformatReads.java:52)\n",
      "java -ea -Xmx195640m -cp /usr/local/bin/bbmap/current/ jgi.FuseSequence in1=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R2_001.fastq.gz in2=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz out=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz overwrite=t interleaved=f fusepairs=t pad=0\n",
      "Executing jgi.FuseSequence [in1=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R2_001.fastq.gz, in2=/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz, out=/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz, overwrite=t, interleaved=f, fusepairs=t, pad=0]\n",
      "\n",
      "Set INTERLEAVED to false\n",
      "pigz: skipping: /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R2_001.fastq.gz does not exist\n",
      "Input is being processed as paired\n",
      "Writing interleaved.pigz: skipping: /media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz does not exist\n",
      "\n",
      "Time:                         \t0.152 seconds.\n",
      "Reads Processed:           0 \t0.00k reads/sec\n",
      "Bases Processed:           0 \t0.00m bases/sec\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-e70cbea8f3f6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m[\u001b[0m\u001b[0mfuseBCs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msample_id\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamples_list\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-7-e70cbea8f3f6>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;34m[\u001b[0m\u001b[0mfuseBCs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msample_id\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamples_list\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-6-0673c4d1bf5e>\u001b[0m in \u001b[0;36mfuseBCs\u001b[0;34m(sample_id)\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0mfuse_call\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0;34m\"/usr/local/bin/bbmap/fuse.sh in1='\"\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0minPathR2\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\"' in2='\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mrevCompR1\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"' out='\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0moutPath\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"' overwrite=t interleaved=f fusepairs=t pad=0\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'$fuse_call  #BBMap is a splice-aware global aligner'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrevCompR1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfuse_call\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"outPath: \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0moutPath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/media/data/AtteR/scifi-analysis/scifi6/220222_NB502004_0181_AHK7L2AFX3/demultiplexed/SciFi_10X_Pilot6_Pooled/SciFi_Pool1-2_EnrichedBCs_All_RC_RC_R1_001.fastq.gz'"
     ]
    }
   ],
   "source": [
    "[fuseBCs(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "SciFi_Kapa_AAV_aSyn_S6_RC_R1_001.fastq.gz, \n",
    "SciFi_Kapa_AAV_aSyn_S6_R21_001.fastq.gz\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_RC_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_new_cDNA_Pool1_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_old_cDNA_Pool2_R1_001.fastq.gz']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R1*.fastq.gz'))\n",
    "#samples_list= [item for item in samples_list if 'Undetermined' not in item]\n",
    "#samples_list= [item for item in samples_list if 'BC' in item]\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/indata/SciFi_new_cDNA_Pool1_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_old_cDNA_Pool2_R21_001.fastq.gz']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R21*.fastq.gz'))\n",
    "#samples_list= [item for item in samples_list if 'Undetermined' not in item]\n",
    "samples_list= [item for item in samples_list if 'EnrichedBCs' not in item]\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trim polyA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runTrimPA(sample_id) :\n",
    "    print(sample_id)\n",
    "    outPath = export_path + ntpath.basename(sample_id).replace('R21_001.fastq.gz','R21_trimPA.fastq.gz')\n",
    "    samplePair = base_path + ntpath.basename(sample_id).replace('R21_001.fastq.gz','R3_001.fastq.gz')\n",
    "    outPathR3 = export_path + ntpath.basename(samplePair).replace('R3_001.fastq.gz','R3_trimPA.fastq.gz')\n",
    "    cutAdapt_call=\"cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o \"+ outPathR3 + \" -p \" + outPath + \" \" + samplePair + \" \" + sample_id\n",
    "    print(cutAdapt_call)\n",
    "    !$cutAdapt_call "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_new_cDNA_Pool1_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_old_cDNA_Pool2_R21_001.fastq.gz']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#samples_list = sorted(glob.glob(base_path+'*_R21_001.fastq.gz'))\n",
    "#For scifi6\n",
    "samples_list = sorted(glob.glob(base_path+'*_R21_001.fastq.gz'))\n",
    "#samples_list= [item for item in samples_list if 'EnrichedBCs' not in item]\n",
    "samples_list\n",
    "#!pip3 install cutadapt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz\n",
      "cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz\n",
      "This is cutadapt 3.5 with Python 3.8.5\n",
      "Command line parameters: --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a A{30}N{104} -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_R21_001.fastq.gz\n",
      "Processing reads on 48 cores in paired-end mode ...\n",
      "Done           00:07:20    84,309,009 reads @   5.2 µs/read;  11.49 M reads/minute\n",
      "Finished in 440.24 s (5 µs/read; 11.49 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         84,309,009\n",
      "  Read 1 with adapter:              13,533,015 (16.1%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:           2,703,975 (3.2%)\n",
      "Pairs written (passing filters):    81,605,034 (96.8%)\n",
      "\n",
      "Total basepairs processed: 13,320,823,422 bp\n",
      "  Read 1: 9,695,536,035 bp\n",
      "  Read 2: 3,625,287,387 bp\n",
      "Quality-trimmed:             584,937,556 bp (4.4%)\n",
      "  Read 1:   548,624,307 bp\n",
      "  Read 2:    36,313,249 bp\n",
      "Total written (filtered):  12,110,122,586 bp (90.9%)\n",
      "  Read 1: 8,634,906,021 bp\n",
      "  Read 2: 3,475,216,565 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GCTGCACCGTCAGA; Type: regular 5'; Length: 14; Trimmed: 1736 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-14 bp: 1\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t1\t20.1\t1\t0 1\n",
      "12\t48\t5.0\t1\t0 48\n",
      "13\t8\t1.3\t1\t0 8\n",
      "14\t16\t0.3\t1\t6 10\n",
      "15\t8\t0.3\t1\t0 8\n",
      "16\t16\t0.3\t1\t0 16\n",
      "17\t1\t0.3\t1\t0 1\n",
      "18\t1\t0.3\t1\t0 1\n",
      "19\t5\t0.3\t1\t0 5\n",
      "20\t17\t0.3\t1\t0 17\n",
      "21\t25\t0.3\t1\t0 25\n",
      "22\t16\t0.3\t1\t0 16\n",
      "23\t8\t0.3\t1\t0 8\n",
      "24\t5\t0.3\t1\t0 5\n",
      "25\t22\t0.3\t1\t0 22\n",
      "26\t17\t0.3\t1\t0 17\n",
      "27\t36\t0.3\t1\t0 36\n",
      "28\t29\t0.3\t1\t0 29\n",
      "29\t31\t0.3\t1\t0 31\n",
      "30\t9\t0.3\t1\t0 9\n",
      "31\t7\t0.3\t1\t0 7\n",
      "32\t6\t0.3\t1\t0 6\n",
      "33\t7\t0.3\t1\t0 7\n",
      "34\t32\t0.3\t1\t0 32\n",
      "35\t12\t0.3\t1\t0 12\n",
      "36\t18\t0.3\t1\t0 18\n",
      "37\t9\t0.3\t1\t0 9\n",
      "38\t21\t0.3\t1\t0 21\n",
      "39\t36\t0.3\t1\t0 36\n",
      "40\t23\t0.3\t1\t0 23\n",
      "41\t11\t0.3\t1\t0 11\n",
      "42\t8\t0.3\t1\t0 8\n",
      "43\t37\t0.3\t1\t0 37\n",
      "44\t27\t0.3\t1\t0 27\n",
      "45\t39\t0.3\t1\t0 39\n",
      "46\t6\t0.3\t1\t0 6\n",
      "47\t3\t0.3\t1\t0 3\n",
      "48\t12\t0.3\t1\t0 12\n",
      "49\t84\t0.3\t1\t0 84\n",
      "50\t41\t0.3\t1\t0 41\n",
      "51\t16\t0.3\t1\t0 16\n",
      "52\t9\t0.3\t1\t0 9\n",
      "53\t22\t0.3\t1\t0 22\n",
      "54\t8\t0.3\t1\t0 8\n",
      "55\t39\t0.3\t1\t0 39\n",
      "56\t10\t0.3\t1\t0 10\n",
      "57\t4\t0.3\t1\t0 4\n",
      "58\t16\t0.3\t1\t0 16\n",
      "59\t29\t0.3\t1\t0 29\n",
      "60\t47\t0.3\t1\t0 47\n",
      "61\t6\t0.3\t1\t0 6\n",
      "62\t39\t0.3\t1\t0 39\n",
      "63\t3\t0.3\t1\t0 3\n",
      "64\t4\t0.3\t1\t0 4\n",
      "65\t8\t0.3\t1\t0 8\n",
      "66\t30\t0.3\t1\t0 30\n",
      "67\t34\t0.3\t1\t0 34\n",
      "68\t9\t0.3\t1\t0 9\n",
      "69\t17\t0.3\t1\t0 17\n",
      "70\t7\t0.3\t1\t0 7\n",
      "71\t15\t0.3\t1\t0 15\n",
      "72\t55\t0.3\t1\t0 55\n",
      "73\t57\t0.3\t1\t0 57\n",
      "74\t13\t0.3\t1\t0 13\n",
      "75\t8\t0.3\t1\t0 8\n",
      "76\t8\t0.3\t1\t0 8\n",
      "77\t34\t0.3\t1\t0 34\n",
      "78\t3\t0.3\t1\t0 3\n",
      "79\t7\t0.3\t1\t0 7\n",
      "80\t2\t0.3\t1\t0 2\n",
      "81\t4\t0.3\t1\t0 4\n",
      "82\t13\t0.3\t1\t0 13\n",
      "83\t20\t0.3\t1\t0 20\n",
      "84\t6\t0.3\t1\t0 6\n",
      "85\t44\t0.3\t1\t0 44\n",
      "86\t8\t0.3\t1\t0 8\n",
      "87\t44\t0.3\t1\t0 44\n",
      "88\t12\t0.3\t1\t0 12\n",
      "89\t3\t0.3\t1\t0 3\n",
      "90\t41\t0.3\t1\t0 41\n",
      "91\t25\t0.3\t1\t0 25\n",
      "92\t17\t0.3\t1\t0 17\n",
      "93\t8\t0.3\t1\t0 8\n",
      "94\t2\t0.3\t1\t0 2\n",
      "95\t12\t0.3\t1\t0 12\n",
      "96\t6\t0.3\t1\t0 6\n",
      "97\t15\t0.3\t1\t0 15\n",
      "98\t11\t0.3\t1\t0 11\n",
      "99\t4\t0.3\t1\t0 4\n",
      "100\t3\t0.3\t1\t0 3\n",
      "101\t43\t0.3\t1\t0 43\n",
      "102\t2\t0.3\t1\t0 2\n",
      "103\t8\t0.3\t1\t0 8\n",
      "104\t8\t0.3\t1\t0 8\n",
      "105\t14\t0.3\t1\t0 14\n",
      "106\t10\t0.3\t1\t0 10\n",
      "108\t1\t0.3\t1\t0 1\n",
      "110\t3\t0.3\t1\t0 3\n",
      "111\t12\t0.3\t1\t0 12\n",
      "112\t4\t0.3\t1\t0 4\n",
      "113\t5\t0.3\t1\t0 5\n",
      "114\t5\t0.3\t1\t0 5\n",
      "115\t6\t0.3\t1\t0 6\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN; Type: regular 3'; Length: 134; Trimmed: 13531279 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30 bp: 3\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.0%\n",
      "  C: 35.1%\n",
      "  G: 35.2%\n",
      "  T: 29.1%\n",
      "  none/other: 0.6%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t70\t20.1\t1\t0 70\n",
      "12\t400658\t5.0\t1\t161357 239301\n",
      "13\t223807\t1.3\t1\t156278 67529\n",
      "14\t222577\t0.3\t1\t160077 62500\n",
      "15\t205539\t0.1\t1\t148865 56674\n",
      "16\t174520\t0.0\t1\t122517 52003\n",
      "17\t148438\t0.0\t1\t101805 46633\n",
      "18\t141828\t0.0\t1\t104623 37205\n",
      "19\t153091\t0.0\t1\t116843 36248\n",
      "20\t268145\t0.0\t2\t132052 37646 98447\n",
      "21\t294021\t0.0\t2\t152528 40142 101351\n",
      "22\t311441\t0.0\t2\t174634 40579 96228\n",
      "23\t345378\t0.0\t2\t204752 42765 97861\n",
      "24\t393301\t0.0\t2\t243285 52968 97048\n",
      "25\t452082\t0.0\t2\t291567 52098 108417\n",
      "26\t470164\t0.0\t2\t300218 54063 115883\n",
      "27\t527592\t0.0\t2\t343349 55462 128781\n",
      "28\t524277\t0.0\t2\t336323 69987 117967\n",
      "29\t500511\t0.0\t2\t273993 83909 142607 2\n",
      "30\t649548\t0.0\t3\t190509 92639 165258 201142\n",
      "31\t617278\t0.0\t3\t120328 90105 173982 232863\n",
      "32\t582928\t0.0\t3\t76458 79015 164625 262830\n",
      "33\t537372\t0.0\t3\t51354 63298 141412 281308\n",
      "34\t461292\t0.0\t3\t37480 49265 110602 263945\n",
      "35\t406488\t0.0\t3\t30792 41030 87569 247097\n",
      "36\t336815\t0.0\t3\t27973 32703 65721 210418\n",
      "37\t279084\t0.0\t3\t25860 24866 49019 179339\n",
      "38\t249845\t0.0\t3\t24768 19440 39379 166258\n",
      "39\t234515\t0.0\t3\t24421 15599 33194 161301\n",
      "40\t216322\t0.0\t3\t25080 12572 28755 149915\n",
      "41\t197558\t0.0\t3\t26748 11204 25999 133607\n",
      "42\t158958\t0.0\t3\t26475 9674 21204 101605\n",
      "43\t131611\t0.0\t3\t27362 8912 19085 76252\n",
      "44\t107759\t0.0\t3\t26903 7978 16668 56210\n",
      "45\t93649\t0.0\t3\t26729 7739 14943 44238\n",
      "46\t85674\t0.0\t3\t26503 7565 14408 37198\n",
      "47\t80934\t0.0\t3\t26771 7876 13244 33043\n",
      "48\t81612\t0.0\t3\t28962 7891 14199 30560\n",
      "49\t77248\t0.0\t3\t29137 7606 13391 27114\n",
      "50\t76524\t0.0\t3\t29956 7637 13432 25499\n",
      "51\t71389\t0.0\t3\t27891 7196 12805 23497\n",
      "52\t71417\t0.0\t3\t26544 7610 12720 24543\n",
      "53\t70222\t0.0\t3\t27950 6795 13041 22436\n",
      "54\t68430\t0.0\t3\t26931 7254 12489 21756\n",
      "55\t70938\t0.0\t3\t28584 7059 12651 22644\n",
      "56\t66807\t0.0\t3\t27392 6898 12089 20428\n",
      "57\t65470\t0.0\t3\t25395 6754 11849 21472\n",
      "58\t66201\t0.0\t3\t26051 7029 12253 20868\n",
      "59\t66270\t0.0\t3\t26209 6643 11863 21555\n",
      "60\t76056\t0.0\t3\t27455 7271 14110 27220\n",
      "61\t63805\t0.0\t3\t24921 6681 11673 20530\n",
      "62\t60470\t0.0\t3\t23743 6392 10704 19631\n",
      "63\t59837\t0.0\t3\t23263 5976 10603 19995\n",
      "64\t57821\t0.0\t3\t22289 5816 10902 18814\n",
      "65\t58041\t0.0\t3\t21490 5963 11505 19083\n",
      "66\t56760\t0.0\t3\t21689 6243 10791 18037\n",
      "67\t52925\t0.0\t3\t20007 5631 9833 17454\n",
      "68\t51413\t0.0\t3\t19196 5317 10027 16873\n",
      "69\t49579\t0.0\t3\t18434 4937 9592 16616\n",
      "70\t46737\t0.0\t3\t17437 4992 8806 15502\n",
      "71\t46076\t0.0\t3\t17858 4823 8539 14856\n",
      "72\t44799\t0.0\t3\t16325 4486 8615 15373\n",
      "73\t42982\t0.0\t3\t15325 4356 7825 15476\n",
      "74\t39356\t0.0\t3\t13856 4112 7578 13810\n",
      "75\t38924\t0.0\t3\t13454 4126 7110 14234\n",
      "76\t37152\t0.0\t3\t12944 3917 6942 13349\n",
      "77\t35509\t0.0\t3\t12098 3784 6979 12648\n",
      "78\t33371\t0.0\t3\t11607 3498 6606 11660\n",
      "79\t31376\t0.0\t3\t10861 3021 6077 11417\n",
      "80\t31316\t0.0\t3\t10378 3041 6785 11112\n",
      "81\t28246\t0.0\t3\t10054 2814 5201 10177\n",
      "82\t28173\t0.0\t3\t9199 2937 5494 10543\n",
      "83\t27103\t0.0\t3\t8557 3220 5378 9948\n",
      "84\t23522\t0.0\t3\t7546 2435 4549 8992\n",
      "85\t22301\t0.0\t3\t6802 2458 4745 8296\n",
      "86\t20516\t0.0\t3\t6651 2131 4143 7591\n",
      "87\t21346\t0.0\t3\t6611 2119 4307 8309\n",
      "88\t19737\t0.0\t3\t6454 2162 3758 7363\n",
      "89\t19689\t0.0\t3\t6744 1984 3791 7170\n",
      "90\t21017\t0.0\t3\t7610 2078 3891 7438\n",
      "91\t19401\t0.0\t3\t7783 2085 3440 6093\n",
      "92\t19707\t0.0\t3\t7948 2038 3634 6087\n",
      "93\t20445\t0.0\t3\t7127 2961 3805 6552\n",
      "94\t16177\t0.0\t3\t5589 1743 3312 5533\n",
      "95\t14832\t0.0\t3\t5198 1560 2968 5106\n",
      "96\t14391\t0.0\t3\t4235 1528 3356 5272\n",
      "97\t13822\t0.0\t3\t3707 1437 3047 5631\n",
      "98\t13555\t0.0\t3\t3918 1394 2874 5369\n",
      "99\t12884\t0.0\t3\t3511 1358 2673 5342\n",
      "100\t12557\t0.0\t3\t3592 1256 2595 5114\n",
      "101\t13554\t0.0\t3\t4033 1421 3124 4976\n",
      "102\t10409\t0.0\t3\t3674 897 1786 4052\n",
      "103\t9609\t0.0\t3\t3380 920 1769 3540\n",
      "104\t8846\t0.0\t3\t2155 1025 1869 3797\n",
      "105\t6851\t0.0\t3\t1181 661 1792 3217\n",
      "106\t6768\t0.0\t3\t704 582 1445 4037\n",
      "107\t3361\t0.0\t3\t276 259 809 2017\n",
      "108\t4103\t0.0\t3\t314 286 849 2654\n",
      "109\t3090\t0.0\t3\t303 217 629 1941\n",
      "110\t3463\t0.0\t3\t468 252 735 2008\n",
      "111\t3298\t0.0\t3\t554 250 644 1850\n",
      "112\t1744\t0.0\t3\t385 120 298 941\n",
      "113\t1721\t0.0\t3\t355 250 268 848\n",
      "114\t2784\t0.0\t3\t814 507 624 839\n",
      "115\t10354\t0.0\t3\t5856 1196 1421 1881\n",
      "/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz\n",
      "cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz\n",
      "This is cutadapt 3.5 with Python 3.8.5\n",
      "Command line parameters: --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a A{30}N{104} -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_001.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing reads on 48 cores in paired-end mode ...\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 641, in run\r\n",
      "    raise e\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "ERROR: Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 567, in run\r\n",
      "    with self._opener.xopen(self.path, 'rb') as f:\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 192, in xopen\r\n",
      "    f = open_raise_limit(\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/utils.py\", line 51, in open_raise_limit\r\n",
      "    f = func(*args, **kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 821, in xopen\r\n",
      "    opened_file = _open_gz(filename, mode, compresslevel, threads, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/xopen/__init__.py\", line 695, in _open_gz\r\n",
      "    return igzip.open(filename, mode, **text_mode_kwargs)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 92, in open\r\n",
      "    binary_file = IGzipFile(filename, gz_mode, compresslevel)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/isal/igzip.py\", line 151, in __init__\r\n",
      "    super().__init__(filename, mode, compresslevel, fileobj, mtime)\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/lib/python3.8/gzip.py\", line 173, in __init__\r\n",
      "    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n",
      "\r\n",
      "Traceback (most recent call last):\r\n",
      "  File \"/home/lcadmin/miniconda3/envs/scifi-analysis/bin/cutadapt\", line 10, in <module>\r\n",
      "    sys.exit(main_cli())\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/__main__.py\", line 877, in main_cli\r\n",
      "    main(sys.argv[1:])\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/__main__.py\", line 942, in main\r\n",
      "    stats = r.run()\r\n",
      "  File \"/home/lcadmin/.local/lib/python3.8/site-packages/cutadapt/pipeline.py\", line 841, in run\r\n",
      "    raise e\r\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/tmp/indata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_001.fastq.gz'\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Pool1-2_EnrichedBCs_R21_001.fastq.gz\n",
      "cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_R21_001.fastq.gz\n",
      "This is cutadapt 3.5 with Python 3.8.5\n",
      "Command line parameters: --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a A{30}N{104} -o /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R21_trimPA.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_R3_001.fastq.gz /tmp/indata/SciFi_Pool1-2_EnrichedBCs_R21_001.fastq.gz\n",
      "Processing reads on 48 cores in paired-end mode ...\n",
      "Done           00:01:20    15,845,078 reads @   5.1 µs/read;  11.76 M reads/minute\n",
      "Finished in 80.89 s (5 µs/read; 11.75 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         15,845,078\n",
      "  Read 1 with adapter:                 381,836 (2.4%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:             426,303 (2.7%)\n",
      "Pairs written (passing filters):    15,418,775 (97.3%)\n",
      "\n",
      "Total basepairs processed: 2,503,522,324 bp\n",
      "  Read 1: 1,822,183,970 bp\n",
      "  Read 2:   681,338,354 bp\n",
      "Quality-trimmed:              76,077,264 bp (3.0%)\n",
      "  Read 1:    70,727,224 bp\n",
      "  Read 2:     5,350,040 bp\n",
      "Total written (filtered):  2,392,674,397 bp (95.6%)\n",
      "  Read 1: 1,734,043,337 bp\n",
      "  Read 2:   658,631,060 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GCTGCACCGTCAGA; Type: regular 5'; Length: 14; Trimmed: 44 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-14 bp: 1\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "14\t4\t0.1\t1\t3 1\n",
      "19\t1\t0.1\t1\t0 1\n",
      "20\t1\t0.1\t1\t0 1\n",
      "21\t1\t0.1\t1\t0 1\n",
      "24\t1\t0.1\t1\t0 1\n",
      "28\t1\t0.1\t1\t0 1\n",
      "29\t1\t0.1\t1\t0 1\n",
      "34\t1\t0.1\t1\t0 1\n",
      "39\t2\t0.1\t1\t0 2\n",
      "44\t2\t0.1\t1\t0 2\n",
      "45\t2\t0.1\t1\t0 2\n",
      "49\t3\t0.1\t1\t0 3\n",
      "50\t2\t0.1\t1\t0 2\n",
      "55\t3\t0.1\t1\t0 3\n",
      "60\t2\t0.1\t1\t0 2\n",
      "72\t2\t0.1\t1\t0 2\n",
      "77\t3\t0.1\t1\t0 3\n",
      "83\t2\t0.1\t1\t0 2\n",
      "84\t3\t0.1\t1\t0 3\n",
      "85\t1\t0.1\t1\t0 1\n",
      "87\t1\t0.1\t1\t0 1\n",
      "88\t1\t0.1\t1\t0 1\n",
      "90\t2\t0.1\t1\t0 2\n",
      "92\t1\t0.1\t1\t0 1\n",
      "105\t1\t0.1\t1\t0 1\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN; Type: regular 3'; Length: 134; Trimmed: 381792 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30 bp: 3\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.0%\n",
      "  C: 34.5%\n",
      "  G: 33.2%\n",
      "  T: 30.7%\n",
      "  none/other: 1.6%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t4\t3.8\t1\t0 4\n",
      "12\t11025\t0.9\t1\t4347 6678\n",
      "13\t6096\t0.2\t1\t4098 1998\n",
      "14\t5973\t0.1\t1\t4141 1832\n",
      "15\t5678\t0.0\t1\t4031 1647\n",
      "16\t4778\t0.0\t1\t3328 1450\n",
      "17\t4030\t0.0\t1\t2726 1304\n",
      "18\t3849\t0.0\t1\t2865 984\n",
      "19\t3985\t0.0\t1\t3014 971\n",
      "20\t7319\t0.0\t2\t3464 1028 2827\n",
      "21\t8060\t0.0\t2\t4023 1113 2924\n",
      "22\t8562\t0.0\t2\t4573 1175 2814\n",
      "23\t9260\t0.0\t2\t5191 1221 2848\n",
      "24\t10419\t0.0\t2\t6257 1428 2734\n",
      "25\t11739\t0.0\t2\t7308 1477 2954\n",
      "26\t11962\t0.0\t2\t7369 1463 3130\n",
      "27\t13520\t0.0\t2\t8604 1432 3484\n",
      "28\t13086\t0.0\t2\t8252 1739 3095\n",
      "29\t12800\t0.0\t2\t6918 2058 3823 1\n",
      "30\t17071\t0.0\t3\t5019 2364 4289 5399\n",
      "31\t15936\t0.0\t3\t3082 2259 4485 6110\n",
      "32\t15027\t0.0\t3\t1970 2082 4194 6781\n",
      "33\t13897\t0.0\t3\t1336 1618 3768 7175\n",
      "34\t12093\t0.0\t3\t950 1315 2989 6839\n",
      "35\t10728\t0.0\t3\t787 1084 2384 6473\n",
      "36\t9199\t0.0\t3\t718 868 1849 5764\n",
      "37\t7402\t0.0\t3\t633 677 1417 4675\n",
      "38\t6754\t0.0\t3\t621 571 1086 4476\n",
      "39\t6313\t0.0\t3\t624 434 978 4277\n",
      "40\t5836\t0.0\t3\t626 366 773 4071\n",
      "41\t5466\t0.0\t3\t619 301 823 3723\n",
      "42\t4209\t0.0\t3\t659 253 565 2732\n",
      "43\t3668\t0.0\t3\t658 252 619 2139\n",
      "44\t3122\t0.0\t3\t670 236 534 1682\n",
      "45\t2651\t0.0\t3\t638 230 481 1302\n",
      "46\t2501\t0.0\t3\t647 218 433 1203\n",
      "47\t2396\t0.0\t3\t670 233 381 1112\n",
      "48\t2405\t0.0\t3\t731 248 436 990\n",
      "49\t2266\t0.0\t3\t674 211 407 974\n",
      "50\t2247\t0.0\t3\t744 239 406 858\n",
      "51\t2086\t0.0\t3\t678 209 379 820\n",
      "52\t2159\t0.0\t3\t680 225 402 852\n",
      "53\t2106\t0.0\t3\t655 202 449 800\n",
      "54\t1963\t0.0\t3\t605 228 384 746\n",
      "55\t2059\t0.0\t3\t684 214 358 803\n",
      "56\t1914\t0.0\t3\t610 188 394 722\n",
      "57\t2001\t0.0\t3\t597 242 390 772\n",
      "58\t1993\t0.0\t3\t638 196 405 754\n",
      "59\t2062\t0.0\t3\t629 206 388 839\n",
      "60\t2294\t0.0\t3\t654 227 475 938\n",
      "61\t1916\t0.0\t3\t570 197 377 772\n",
      "62\t1848\t0.0\t3\t573 167 361 747\n",
      "63\t1855\t0.0\t3\t591 177 351 736\n",
      "64\t1775\t0.0\t3\t554 178 363 680\n",
      "65\t1853\t0.0\t3\t552 183 361 757\n",
      "66\t1704\t0.0\t3\t486 169 342 707\n",
      "67\t1694\t0.0\t3\t463 162 351 718\n",
      "68\t1624\t0.0\t3\t480 165 307 672\n",
      "69\t1594\t0.0\t3\t448 150 340 656\n",
      "70\t1583\t0.0\t3\t424 174 322 663\n",
      "71\t1555\t0.0\t3\t442 148 316 649\n",
      "72\t1504\t0.0\t3\t407 148 292 657\n",
      "73\t1494\t0.0\t3\t392 138 274 690\n",
      "74\t1345\t0.0\t3\t354 145 256 590\n",
      "75\t1407\t0.0\t3\t376 143 299 589\n",
      "76\t1403\t0.0\t3\t338 138 275 652\n",
      "77\t1237\t0.0\t3\t307 122 265 543\n",
      "78\t1272\t0.0\t3\t306 117 291 558\n",
      "79\t1158\t0.0\t3\t279 112 230 537\n",
      "80\t1207\t0.0\t3\t298 135 255 519\n",
      "81\t1086\t0.0\t3\t279 98 212 497\n",
      "82\t1178\t0.0\t3\t256 139 218 565\n",
      "83\t1075\t0.0\t3\t245 127 230 473\n",
      "84\t992\t0.0\t3\t207 102 212 471\n",
      "85\t991\t0.0\t3\t189 115 211 476\n",
      "86\t932\t0.0\t3\t195 94 206 437\n",
      "87\t924\t0.0\t3\t185 78 174 487\n",
      "88\t897\t0.0\t3\t188 97 176 436\n",
      "89\t866\t0.0\t3\t187 69 195 415\n",
      "90\t970\t0.0\t3\t223 105 171 471\n",
      "91\t847\t0.0\t3\t187 93 184 383\n",
      "92\t959\t0.0\t3\t241 96 192 430\n",
      "93\t900\t0.0\t3\t204 102 193 401\n",
      "94\t800\t0.0\t3\t172 79 160 389\n",
      "95\t728\t0.0\t3\t149 72 131 376\n",
      "96\t744\t0.0\t3\t126 74 170 374\n",
      "97\t742\t0.0\t3\t141 74 147 380\n",
      "98\t720\t0.0\t3\t131 54 165 370\n",
      "99\t767\t0.0\t3\t130 65 171 401\n",
      "100\t737\t0.0\t3\t139 69 153 376\n",
      "101\t793\t0.0\t3\t140 78 181 394\n",
      "102\t668\t0.0\t3\t133 66 127 342\n",
      "103\t653\t0.0\t3\t155 69 134 295\n",
      "104\t675\t0.0\t3\t113 60 128 374\n",
      "105\t547\t0.0\t3\t97 52 124 274\n",
      "106\t547\t0.0\t3\t58 41 106 342\n",
      "107\t318\t0.0\t3\t36 38 60 184\n",
      "108\t378\t0.0\t3\t42 24 77 235\n",
      "109\t306\t0.0\t3\t47 30 50 179\n",
      "110\t365\t0.0\t3\t61 32 79 193\n",
      "111\t392\t0.0\t3\t69 32 70 221\n",
      "112\t304\t0.0\t3\t68 21 54 161\n",
      "113\t343\t0.0\t3\t84 43 71 145\n",
      "114\t789\t0.0\t3\t361 95 129 204\n",
      "115\t1862\t0.0\t3\t1221 161 197 283\n",
      "/tmp/indata/SciFi_new_cDNA_Pool1_R21_001.fastq.gz\n",
      "cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o /tmp/outdata/SciFi_new_cDNA_Pool1_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz /tmp/indata/SciFi_new_cDNA_Pool1_R3_001.fastq.gz /tmp/indata/SciFi_new_cDNA_Pool1_R21_001.fastq.gz\n",
      "This is cutadapt 3.5 with Python 3.8.5\n",
      "Command line parameters: --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a A{30}N{104} -o /tmp/outdata/SciFi_new_cDNA_Pool1_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz /tmp/indata/SciFi_new_cDNA_Pool1_R3_001.fastq.gz /tmp/indata/SciFi_new_cDNA_Pool1_R21_001.fastq.gz\n",
      "Processing reads on 48 cores in paired-end mode ...\n",
      "Done           00:04:15    48,677,522 reads @   5.3 µs/read;  11.42 M reads/minute\n",
      "Finished in 255.86 s (5 µs/read; 11.42 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         48,677,522\n",
      "  Read 1 with adapter:               8,752,929 (18.0%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:           1,496,319 (3.1%)\n",
      "Pairs written (passing filters):    47,181,203 (96.9%)\n",
      "\n",
      "Total basepairs processed: 7,691,048,476 bp\n",
      "  Read 1: 5,597,915,030 bp\n",
      "  Read 2: 2,093,133,446 bp\n",
      "Quality-trimmed:             350,490,044 bp (4.6%)\n",
      "  Read 1:   320,474,487 bp\n",
      "  Read 2:    30,015,557 bp\n",
      "Total written (filtered):  6,950,928,856 bp (90.4%)\n",
      "  Read 1: 4,950,780,652 bp\n",
      "  Read 2: 2,000,148,204 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GCTGCACCGTCAGA; Type: regular 5'; Length: 14; Trimmed: 1400 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-14 bp: 1\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "12\t42\t2.9\t1\t0 42\n",
      "13\t4\t0.7\t1\t0 4\n",
      "14\t7\t0.2\t1\t0 7\n",
      "15\t4\t0.2\t1\t0 4\n",
      "16\t16\t0.2\t1\t0 16\n",
      "17\t1\t0.2\t1\t0 1\n",
      "19\t4\t0.2\t1\t0 4\n",
      "20\t9\t0.2\t1\t0 9\n",
      "21\t23\t0.2\t1\t0 23\n",
      "22\t11\t0.2\t1\t0 11\n",
      "23\t4\t0.2\t1\t0 4\n",
      "24\t4\t0.2\t1\t0 4\n",
      "25\t19\t0.2\t1\t0 19\n",
      "26\t17\t0.2\t1\t0 17\n",
      "27\t24\t0.2\t1\t0 24\n",
      "28\t22\t0.2\t1\t0 22\n",
      "29\t26\t0.2\t1\t0 26\n",
      "30\t7\t0.2\t1\t0 7\n",
      "31\t4\t0.2\t1\t0 4\n",
      "32\t5\t0.2\t1\t0 5\n",
      "33\t6\t0.2\t1\t0 6\n",
      "34\t30\t0.2\t1\t0 30\n",
      "35\t12\t0.2\t1\t0 12\n",
      "36\t12\t0.2\t1\t0 12\n",
      "37\t8\t0.2\t1\t0 8\n",
      "38\t11\t0.2\t1\t0 11\n",
      "39\t28\t0.2\t1\t0 28\n",
      "40\t21\t0.2\t1\t0 21\n",
      "41\t6\t0.2\t1\t0 6\n",
      "42\t3\t0.2\t1\t0 3\n",
      "43\t36\t0.2\t1\t0 36\n",
      "44\t13\t0.2\t1\t0 13\n",
      "45\t28\t0.2\t1\t0 28\n",
      "46\t5\t0.2\t1\t0 5\n",
      "47\t3\t0.2\t1\t0 3\n",
      "48\t11\t0.2\t1\t0 11\n",
      "49\t71\t0.2\t1\t0 71\n",
      "50\t29\t0.2\t1\t0 29\n",
      "51\t14\t0.2\t1\t0 14\n",
      "52\t7\t0.2\t1\t0 7\n",
      "53\t20\t0.2\t1\t0 20\n",
      "54\t7\t0.2\t1\t0 7\n",
      "55\t33\t0.2\t1\t0 33\n",
      "56\t9\t0.2\t1\t0 9\n",
      "57\t4\t0.2\t1\t0 4\n",
      "58\t13\t0.2\t1\t0 13\n",
      "59\t19\t0.2\t1\t0 19\n",
      "60\t35\t0.2\t1\t0 35\n",
      "61\t5\t0.2\t1\t0 5\n",
      "62\t28\t0.2\t1\t0 28\n",
      "63\t3\t0.2\t1\t0 3\n",
      "64\t3\t0.2\t1\t0 3\n",
      "65\t4\t0.2\t1\t0 4\n",
      "66\t24\t0.2\t1\t0 24\n",
      "67\t28\t0.2\t1\t0 28\n",
      "68\t7\t0.2\t1\t0 7\n",
      "69\t16\t0.2\t1\t0 16\n",
      "70\t5\t0.2\t1\t0 5\n",
      "71\t13\t0.2\t1\t0 13\n",
      "72\t43\t0.2\t1\t0 43\n",
      "73\t51\t0.2\t1\t0 51\n",
      "74\t11\t0.2\t1\t0 11\n",
      "75\t5\t0.2\t1\t0 5\n",
      "76\t5\t0.2\t1\t0 5\n",
      "77\t31\t0.2\t1\t0 31\n",
      "78\t3\t0.2\t1\t0 3\n",
      "79\t6\t0.2\t1\t0 6\n",
      "80\t2\t0.2\t1\t0 2\n",
      "81\t3\t0.2\t1\t0 3\n",
      "82\t12\t0.2\t1\t0 12\n",
      "83\t14\t0.2\t1\t0 14\n",
      "84\t3\t0.2\t1\t0 3\n",
      "85\t42\t0.2\t1\t0 42\n",
      "86\t7\t0.2\t1\t0 7\n",
      "87\t41\t0.2\t1\t0 41\n",
      "88\t10\t0.2\t1\t0 10\n",
      "89\t1\t0.2\t1\t0 1\n",
      "90\t39\t0.2\t1\t0 39\n",
      "91\t21\t0.2\t1\t0 21\n",
      "92\t15\t0.2\t1\t0 15\n",
      "93\t7\t0.2\t1\t0 7\n",
      "94\t2\t0.2\t1\t0 2\n",
      "95\t10\t0.2\t1\t0 10\n",
      "96\t4\t0.2\t1\t0 4\n",
      "97\t15\t0.2\t1\t0 15\n",
      "98\t11\t0.2\t1\t0 11\n",
      "99\t2\t0.2\t1\t0 2\n",
      "100\t2\t0.2\t1\t0 2\n",
      "101\t34\t0.2\t1\t0 34\n",
      "102\t1\t0.2\t1\t0 1\n",
      "103\t7\t0.2\t1\t0 7\n",
      "104\t7\t0.2\t1\t0 7\n",
      "105\t13\t0.2\t1\t0 13\n",
      "106\t9\t0.2\t1\t0 9\n",
      "108\t1\t0.2\t1\t0 1\n",
      "110\t1\t0.2\t1\t0 1\n",
      "111\t10\t0.2\t1\t0 10\n",
      "112\t2\t0.2\t1\t0 2\n",
      "113\t4\t0.2\t1\t0 4\n",
      "114\t5\t0.2\t1\t0 5\n",
      "115\t5\t0.2\t1\t0 5\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN; Type: regular 3'; Length: 134; Trimmed: 8751529 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30 bp: 3\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.0%\n",
      "  C: 35.4%\n",
      "  G: 35.7%\n",
      "  T: 28.2%\n",
      "  none/other: 0.6%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t46\t11.6\t1\t0 46\n",
      "12\t265443\t2.9\t1\t106156 159287\n",
      "13\t147180\t0.7\t1\t102915 44265\n",
      "14\t144951\t0.2\t1\t104645 40306\n",
      "15\t133930\t0.0\t1\t97646 36284\n",
      "16\t113685\t0.0\t1\t80315 33370\n",
      "17\t95742\t0.0\t1\t66172 29570\n",
      "18\t91007\t0.0\t1\t67524 23483\n",
      "19\t97982\t0.0\t1\t75192 22790\n",
      "20\t172724\t0.0\t2\t85340 23780 63604\n",
      "21\t187569\t0.0\t2\t96957 25434 65178\n",
      "22\t198128\t0.0\t2\t111675 25341 61112\n",
      "23\t218508\t0.0\t2\t130078 26497 61933\n",
      "24\t248577\t0.0\t2\t155450 31799 61328\n",
      "25\t289005\t0.0\t2\t188363 31690 68952\n",
      "26\t304084\t0.0\t2\t198133 32991 72960\n",
      "27\t347588\t0.0\t2\t232411 33424 81753\n",
      "28\t353006\t0.0\t2\t233870 42885 76251\n",
      "29\t339717\t0.0\t2\t194975 51732 93009 1\n",
      "30\t432125\t0.0\t3\t136739 58511 109668 127207\n",
      "31\t411473\t0.0\t3\t85728 58674 117618 149453\n",
      "32\t390614\t0.0\t3\t53059 51989 113161 172405\n",
      "33\t361322\t0.0\t3\t34347 42537 97328 187110\n",
      "34\t310206\t0.0\t3\t24007 33179 76333 176687\n",
      "35\t272639\t0.0\t3\t19023 27582 59826 166208\n",
      "36\t223602\t0.0\t3\t16812 21991 44155 140644\n",
      "37\t183638\t0.0\t3\t15179 16826 32540 119093\n",
      "38\t163577\t0.0\t3\t14246 13066 26098 110167\n",
      "39\t153703\t0.0\t3\t13892 10254 21908 107649\n",
      "40\t141631\t0.0\t3\t14206 8117 18615 100693\n",
      "41\t129654\t0.0\t3\t15266 7092 16709 90587\n",
      "42\t103455\t0.0\t3\t15194 6037 13675 68549\n",
      "43\t84031\t0.0\t3\t15611 5404 11959 51057\n",
      "44\t67035\t0.0\t3\t15053 4692 10156 37134\n",
      "45\t57402\t0.0\t3\t15146 4588 8935 28733\n",
      "46\t51310\t0.0\t3\t14914 4375 8584 23437\n",
      "47\t47755\t0.0\t3\t15118 4437 7844 20356\n",
      "48\t47568\t0.0\t3\t16286 4462 8353 18467\n",
      "49\t44977\t0.0\t3\t16745 4330 7752 16150\n",
      "50\t44139\t0.0\t3\t16944 4281 7824 15090\n",
      "51\t40701\t0.0\t3\t15748 4096 7178 13679\n",
      "52\t40563\t0.0\t3\t14899 4351 7197 14116\n",
      "53\t40061\t0.0\t3\t15862 3848 7429 12922\n",
      "54\t39098\t0.0\t3\t15481 4110 7083 12424\n",
      "55\t40799\t0.0\t3\t16395 4039 7259 13106\n",
      "56\t38123\t0.0\t3\t15701 3933 6867 11622\n",
      "57\t37177\t0.0\t3\t14571 3774 6685 12147\n",
      "58\t38011\t0.0\t3\t15002 4010 6995 12004\n",
      "59\t38101\t0.0\t3\t15174 3856 6676 12395\n",
      "60\t44494\t0.0\t3\t16322 4199 8101 15872\n",
      "61\t36848\t0.0\t3\t14471 3848 6685 11844\n",
      "62\t34749\t0.0\t3\t13847 3645 6097 11160\n",
      "63\t34795\t0.0\t3\t13750 3447 6058 11540\n",
      "64\t33454\t0.0\t3\t13139 3307 6238 10770\n",
      "65\t33844\t0.0\t3\t12687 3510 6627 11020\n",
      "66\t33505\t0.0\t3\t12914 3698 6308 10585\n",
      "67\t31488\t0.0\t3\t12237 3346 5765 10140\n",
      "68\t30824\t0.0\t3\t11781 3172 5914 9957\n",
      "69\t29832\t0.0\t3\t11417 2923 5726 9766\n",
      "70\t28308\t0.0\t3\t10908 3028 5239 9133\n",
      "71\t28284\t0.0\t3\t11333 3002 5110 8839\n",
      "72\t27495\t0.0\t3\t10353 2685 5214 9243\n",
      "73\t27051\t0.0\t3\t9869 2757 4907 9518\n",
      "74\t24690\t0.0\t3\t8905 2612 4744 8429\n",
      "75\t24484\t0.0\t3\t8646 2503 4411 8924\n",
      "76\t23339\t0.0\t3\t8469 2427 4362 8081\n",
      "77\t22450\t0.0\t3\t7843 2363 4328 7916\n",
      "78\t21332\t0.0\t3\t7642 2196 4161 7333\n",
      "79\t20068\t0.0\t3\t7135 1961 3856 7116\n",
      "80\t20141\t0.0\t3\t6918 1958 4367 6898\n",
      "81\t18481\t0.0\t3\t6829 1809 3372 6471\n",
      "82\t18247\t0.0\t3\t6163 1903 3434 6747\n",
      "83\t17821\t0.0\t3\t5855 2056 3566 6344\n",
      "84\t15599\t0.0\t3\t5182 1602 2970 5845\n",
      "85\t14795\t0.0\t3\t4671 1643 3109 5372\n",
      "86\t13539\t0.0\t3\t4455 1456 2681 4947\n",
      "87\t14413\t0.0\t3\t4544 1402 2923 5544\n",
      "88\t13182\t0.0\t3\t4437 1431 2509 4805\n",
      "89\t13220\t0.0\t3\t4611 1333 2575 4701\n",
      "90\t14225\t0.0\t3\t5277 1393 2619 4936\n",
      "91\t13094\t0.0\t3\t5307 1434 2318 4035\n",
      "92\t13208\t0.0\t3\t5345 1392 2444 4027\n",
      "93\t13452\t0.0\t3\t4840 1920 2529 4163\n",
      "94\t10827\t0.0\t3\t3873 1183 2143 3628\n",
      "95\t9966\t0.0\t3\t3619 1010 1980 3357\n",
      "96\t9628\t0.0\t3\t2893 1037 2252 3446\n",
      "97\t9208\t0.0\t3\t2511 967 2042 3688\n",
      "98\t9023\t0.0\t3\t2706 965 1882 3470\n",
      "99\t8507\t0.0\t3\t2425 899 1719 3464\n",
      "100\t8350\t0.0\t3\t2480 834 1720 3316\n",
      "101\t9104\t0.0\t3\t2801 952 2096 3255\n",
      "102\t6882\t0.0\t3\t2556 595 1149 2582\n",
      "103\t6404\t0.0\t3\t2349 628 1173 2254\n",
      "104\t5815\t0.0\t3\t1497 695 1218 2405\n",
      "105\t4501\t0.0\t3\t793 446 1216 2046\n",
      "106\t4402\t0.0\t3\t422 397 950 2633\n",
      "107\t2098\t0.0\t3\t167 156 521 1254\n",
      "108\t2582\t0.0\t3\t190 175 529 1688\n",
      "109\t1943\t0.0\t3\t182 132 411 1218\n",
      "110\t2237\t0.0\t3\t307 163 461 1306\n",
      "111\t2050\t0.0\t3\t360 157 384 1149\n",
      "112\t1003\t0.0\t3\t219 68 183 533\n",
      "113\t996\t0.0\t3\t189 143 145 519\n",
      "114\t1420\t0.0\t3\t326 307 346 441\n",
      "115\t6470\t0.0\t3\t3650 758 890 1172\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_old_cDNA_Pool2_R21_001.fastq.gz\n",
      "cutadapt --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a 'A{30}N{104}' -o /tmp/outdata/SciFi_old_cDNA_Pool2_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_old_cDNA_Pool2_R21_trimPA.fastq.gz /tmp/indata/SciFi_old_cDNA_Pool2_R3_001.fastq.gz /tmp/indata/SciFi_old_cDNA_Pool2_R21_001.fastq.gz\n",
      "This is cutadapt 3.5 with Python 3.8.5\n",
      "Command line parameters: --cores=0 --nextseq-trim=25 --minimum-length 25 -g GCTGCACCGTCAGA -O 12 -a A{30}N{104} -o /tmp/outdata/SciFi_old_cDNA_Pool2_R3_trimPA.fastq.gz -p /tmp/outdata/SciFi_old_cDNA_Pool2_R21_trimPA.fastq.gz /tmp/indata/SciFi_old_cDNA_Pool2_R3_001.fastq.gz /tmp/indata/SciFi_old_cDNA_Pool2_R21_001.fastq.gz\n",
      "Processing reads on 48 cores in paired-end mode ...\n",
      "Done           00:01:39    19,786,409 reads @   5.0 µs/read;  11.96 M reads/minute\n",
      "Finished in 99.28 s (5 µs/read; 11.96 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         19,786,409\n",
      "  Read 1 with adapter:               4,398,250 (22.2%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:             781,353 (3.9%)\n",
      "Pairs written (passing filters):    19,005,056 (96.1%)\n",
      "\n",
      "Total basepairs processed: 3,126,252,622 bp\n",
      "  Read 1: 2,275,437,035 bp\n",
      "  Read 2:   850,815,587 bp\n",
      "Quality-trimmed:             158,370,248 bp (5.1%)\n",
      "  Read 1:   157,422,596 bp\n",
      "  Read 2:       947,652 bp\n",
      "Total written (filtered):  2,766,519,333 bp (88.5%)\n",
      "  Read 1: 1,950,082,032 bp\n",
      "  Read 2:   816,437,301 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GCTGCACCGTCAGA; Type: regular 5'; Length: 14; Trimmed: 292 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-14 bp: 1\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t1\t4.7\t1\t0 1\n",
      "12\t6\t1.2\t1\t0 6\n",
      "13\t4\t0.3\t1\t0 4\n",
      "14\t5\t0.1\t1\t3 2\n",
      "15\t4\t0.1\t1\t0 4\n",
      "18\t1\t0.1\t1\t0 1\n",
      "20\t7\t0.1\t1\t0 7\n",
      "21\t1\t0.1\t1\t0 1\n",
      "22\t5\t0.1\t1\t0 5\n",
      "23\t4\t0.1\t1\t0 4\n",
      "25\t3\t0.1\t1\t0 3\n",
      "27\t12\t0.1\t1\t0 12\n",
      "28\t6\t0.1\t1\t0 6\n",
      "29\t4\t0.1\t1\t0 4\n",
      "30\t2\t0.1\t1\t0 2\n",
      "31\t3\t0.1\t1\t0 3\n",
      "32\t1\t0.1\t1\t0 1\n",
      "33\t1\t0.1\t1\t0 1\n",
      "34\t1\t0.1\t1\t0 1\n",
      "36\t6\t0.1\t1\t0 6\n",
      "37\t1\t0.1\t1\t0 1\n",
      "38\t10\t0.1\t1\t0 10\n",
      "39\t6\t0.1\t1\t0 6\n",
      "40\t2\t0.1\t1\t0 2\n",
      "41\t5\t0.1\t1\t0 5\n",
      "42\t5\t0.1\t1\t0 5\n",
      "43\t1\t0.1\t1\t0 1\n",
      "44\t12\t0.1\t1\t0 12\n",
      "45\t9\t0.1\t1\t0 9\n",
      "46\t1\t0.1\t1\t0 1\n",
      "48\t1\t0.1\t1\t0 1\n",
      "49\t10\t0.1\t1\t0 10\n",
      "50\t10\t0.1\t1\t0 10\n",
      "51\t2\t0.1\t1\t0 2\n",
      "52\t2\t0.1\t1\t0 2\n",
      "53\t2\t0.1\t1\t0 2\n",
      "54\t1\t0.1\t1\t0 1\n",
      "55\t3\t0.1\t1\t0 3\n",
      "56\t1\t0.1\t1\t0 1\n",
      "58\t3\t0.1\t1\t0 3\n",
      "59\t10\t0.1\t1\t0 10\n",
      "60\t10\t0.1\t1\t0 10\n",
      "61\t1\t0.1\t1\t0 1\n",
      "62\t11\t0.1\t1\t0 11\n",
      "64\t1\t0.1\t1\t0 1\n",
      "65\t4\t0.1\t1\t0 4\n",
      "66\t6\t0.1\t1\t0 6\n",
      "67\t6\t0.1\t1\t0 6\n",
      "68\t2\t0.1\t1\t0 2\n",
      "69\t1\t0.1\t1\t0 1\n",
      "70\t2\t0.1\t1\t0 2\n",
      "71\t2\t0.1\t1\t0 2\n",
      "72\t10\t0.1\t1\t0 10\n",
      "73\t6\t0.1\t1\t0 6\n",
      "74\t2\t0.1\t1\t0 2\n",
      "75\t3\t0.1\t1\t0 3\n",
      "76\t3\t0.1\t1\t0 3\n",
      "79\t1\t0.1\t1\t0 1\n",
      "81\t1\t0.1\t1\t0 1\n",
      "82\t1\t0.1\t1\t0 1\n",
      "83\t4\t0.1\t1\t0 4\n",
      "85\t1\t0.1\t1\t0 1\n",
      "86\t1\t0.1\t1\t0 1\n",
      "87\t2\t0.1\t1\t0 2\n",
      "88\t1\t0.1\t1\t0 1\n",
      "89\t2\t0.1\t1\t0 2\n",
      "91\t4\t0.1\t1\t0 4\n",
      "92\t1\t0.1\t1\t0 1\n",
      "93\t1\t0.1\t1\t0 1\n",
      "95\t2\t0.1\t1\t0 2\n",
      "96\t2\t0.1\t1\t0 2\n",
      "99\t2\t0.1\t1\t0 2\n",
      "100\t1\t0.1\t1\t0 1\n",
      "101\t9\t0.1\t1\t0 9\n",
      "102\t1\t0.1\t1\t0 1\n",
      "103\t1\t0.1\t1\t0 1\n",
      "104\t1\t0.1\t1\t0 1\n",
      "106\t1\t0.1\t1\t0 1\n",
      "110\t2\t0.1\t1\t0 2\n",
      "111\t2\t0.1\t1\t0 2\n",
      "112\t2\t0.1\t1\t0 2\n",
      "113\t1\t0.1\t1\t0 1\n",
      "115\t1\t0.1\t1\t0 1\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN; Type: regular 3'; Length: 134; Trimmed: 4397958 times\n",
      "\n",
      "Minimum overlap: 12\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30 bp: 3\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.0%\n",
      "  C: 34.4%\n",
      "  G: 34.4%\n",
      "  T: 30.7%\n",
      "  none/other: 0.6%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "11\t20\t4.7\t1\t0 20\n",
      "12\t124190\t1.2\t1\t50854 73336\n",
      "13\t70531\t0.3\t1\t49265 21266\n",
      "14\t71653\t0.1\t1\t51291 20362\n",
      "15\t65931\t0.0\t1\t47188 18743\n",
      "16\t56057\t0.0\t1\t38874 17183\n",
      "17\t48666\t0.0\t1\t32907 15759\n",
      "18\t46972\t0.0\t1\t34234 12738\n",
      "19\t51124\t0.0\t1\t38637 12487\n",
      "20\t88102\t0.0\t2\t43248 12838 32016\n",
      "21\t98392\t0.0\t2\t51548 13595 33249\n",
      "22\t104751\t0.0\t2\t58386 14063 32302\n",
      "23\t117610\t0.0\t2\t69483 15047 33080\n",
      "24\t134305\t0.0\t2\t81578 19741 32986\n",
      "25\t151338\t0.0\t2\t95896 18931 36511\n",
      "26\t154118\t0.0\t2\t94716 19609 39793\n",
      "27\t166484\t0.0\t2\t102334 20606 43544\n",
      "28\t158185\t0.0\t2\t94201 25363 38621\n",
      "29\t147994\t0.0\t2\t72100 30119 45775\n",
      "30\t200352\t0.0\t3\t48751 31764 51301 68536\n",
      "31\t189869\t0.0\t3\t31518 29172 51879 77300\n",
      "32\t177287\t0.0\t3\t21429 24944 47270 83644\n",
      "33\t162153\t0.0\t3\t15671 19143 40316 87023\n",
      "34\t138993\t0.0\t3\t12523 14771 31280 80419\n",
      "35\t123121\t0.0\t3\t10982 12364 25359 74416\n",
      "36\t104014\t0.0\t3\t10443 9844 19717 64010\n",
      "37\t88044\t0.0\t3\t10048 7363 15062 55571\n",
      "38\t79514\t0.0\t3\t9901 5803 12195 51615\n",
      "39\t74499\t0.0\t3\t9905 4911 10308 49375\n",
      "40\t68855\t0.0\t3\t10248 4089 9367 45151\n",
      "41\t62438\t0.0\t3\t10863 3811 8467 39297\n",
      "42\t51294\t0.0\t3\t10622 3384 6964 30324\n",
      "43\t43912\t0.0\t3\t11093 3256 6507 23056\n",
      "44\t37602\t0.0\t3\t11180 3050 5978 17394\n",
      "45\t33596\t0.0\t3\t10945 2921 5527 14203\n",
      "46\t31863\t0.0\t3\t10942 2972 5391 12558\n",
      "47\t30783\t0.0\t3\t10983 3206 5019 11575\n",
      "48\t31639\t0.0\t3\t11945 3181 5410 11103\n",
      "49\t30005\t0.0\t3\t11718 3065 5232 9990\n",
      "50\t30138\t0.0\t3\t12268 3117 5202 9551\n",
      "51\t28602\t0.0\t3\t11465 2891 5248 8998\n",
      "52\t28695\t0.0\t3\t10965 3034 5121 9575\n",
      "53\t28055\t0.0\t3\t11433 2745 5163 8714\n",
      "54\t27369\t0.0\t3\t10845 2916 5022 8586\n",
      "55\t28080\t0.0\t3\t11505 2806 5034 8735\n",
      "56\t26770\t0.0\t3\t11081 2777 4828 8084\n",
      "57\t26292\t0.0\t3\t10227 2738 4774 8553\n",
      "58\t26197\t0.0\t3\t10411 2823 4853 8110\n",
      "59\t26107\t0.0\t3\t10406 2581 4799 8321\n",
      "60\t29268\t0.0\t3\t10479 2845 5534 10410\n",
      "61\t25041\t0.0\t3\t9880 2636 4611 7914\n",
      "62\t23873\t0.0\t3\t9323 2580 4246 7724\n",
      "63\t23187\t0.0\t3\t8922 2352 4194 7719\n",
      "64\t22592\t0.0\t3\t8596 2331 4301 7364\n",
      "65\t22344\t0.0\t3\t8251 2270 4517 7306\n",
      "66\t21551\t0.0\t3\t8289 2376 4141 6745\n",
      "67\t19743\t0.0\t3\t7307 2123 3717 6596\n",
      "68\t18965\t0.0\t3\t6935 1980 3806 6244\n",
      "69\t18153\t0.0\t3\t6569 1864 3526 6194\n",
      "70\t16846\t0.0\t3\t6105 1790 3245 5706\n",
      "71\t16237\t0.0\t3\t6083 1673 3113 5368\n",
      "72\t15800\t0.0\t3\t5565 1653 3109 5473\n",
      "73\t14437\t0.0\t3\t5064 1461 2644 5268\n",
      "74\t13321\t0.0\t3\t4597 1355 2578 4791\n",
      "75\t13033\t0.0\t3\t4432 1480 2400 4721\n",
      "76\t12410\t0.0\t3\t4137 1352 2305 4616\n",
      "77\t11822\t0.0\t3\t3948 1299 2386 4189\n",
      "78\t10767\t0.0\t3\t3659 1185 2154 3769\n",
      "79\t10150\t0.0\t3\t3447 948 1991 3764\n",
      "80\t9968\t0.0\t3\t3162 948 2163 3695\n",
      "81\t8679\t0.0\t3\t2946 907 1617 3209\n",
      "82\t8748\t0.0\t3\t2780 895 1842 3231\n",
      "83\t8207\t0.0\t3\t2457 1037 1582 3131\n",
      "84\t6931\t0.0\t3\t2157 731 1367 2676\n",
      "85\t6515\t0.0\t3\t1942 700 1425 2448\n",
      "86\t6045\t0.0\t3\t2001 581 1256 2207\n",
      "87\t6009\t0.0\t3\t1882 639 1210 2278\n",
      "88\t5658\t0.0\t3\t1829 634 1073 2122\n",
      "89\t5603\t0.0\t3\t1946 582 1021 2054\n",
      "90\t5822\t0.0\t3\t2110 580 1101 2031\n",
      "91\t5460\t0.0\t3\t2289 558 938 1675\n",
      "92\t5540\t0.0\t3\t2362 550 998 1630\n",
      "93\t6093\t0.0\t3\t2083 939 1083 1988\n",
      "94\t4550\t0.0\t3\t1544 481 1009 1516\n",
      "95\t4138\t0.0\t3\t1430 478 857 1373\n",
      "96\t4019\t0.0\t3\t1216 417 934 1452\n",
      "97\t3872\t0.0\t3\t1055 396 858 1563\n",
      "98\t3812\t0.0\t3\t1081 375 827 1529\n",
      "99\t3610\t0.0\t3\t956 394 783 1477\n",
      "100\t3470\t0.0\t3\t973 353 722 1422\n",
      "101\t3657\t0.0\t3\t1092 391 847 1327\n",
      "102\t2859\t0.0\t3\t985 236 510 1128\n",
      "103\t2552\t0.0\t3\t876 223 462 991\n",
      "104\t2356\t0.0\t3\t545 270 523 1018\n",
      "105\t1803\t0.0\t3\t291 163 452 897\n",
      "106\t1819\t0.0\t3\t224 144 389 1062\n",
      "107\t945\t0.0\t3\t73 65 228 579\n",
      "108\t1143\t0.0\t3\t82 87 243 731\n",
      "109\t841\t0.0\t3\t74 55 168 544\n",
      "110\t861\t0.0\t3\t100 57 195 509\n",
      "111\t856\t0.0\t3\t125 61 190 480\n",
      "112\t437\t0.0\t3\t98 31 61 247\n",
      "113\t382\t0.0\t3\t82 64 52 184\n",
      "114\t575\t0.0\t3\t127 105 149 194\n",
      "115\t2022\t0.0\t3\t985 277 334 426\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None, None]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#earlier was samples_list[2:] and initially we had enriched bcs (3 files so one of them got included into the loop)in\n",
    "#[runTrimPA(sample_id) for sample_id in samples_list[2:]]\n",
    "[runTrimPA(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R3_trimPA.fastq.gz', '/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R3_trimPA.fastq.gz', '/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R3_trimPA.fastq.gz', '/tmp/outdata/SciFi_new_cDNA_Pool1_R3_trimPA.fastq.gz', '/tmp/outdata/SciFi_old_cDNA_Pool2_R3_trimPA.fastq.gz']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n#copy the file from temp dir for later use as a security check\\nsave_path = output_data2 + \"/trimmed\"\\nfor trimd_file in samples_list:\\n    target_path = save_path + \"/\" + ntpath.basename(trimd_file)\\n    print(target_path)\\n    shutil.copyfile(trimd_file, target_path)\\n'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(export_path+'*R3_trimPA.fastq.gz'))\n",
    "\n",
    "print(samples_list)\n",
    "\n",
    "samples_list = sorted(glob.glob(export_path+'*_trimPA.fastq.gz'))\n",
    "'''\n",
    "#copy the file from temp dir for later use as a security check\n",
    "save_path = output_data2 + \"/trimmed\"\n",
    "for trimd_file in samples_list:\n",
    "    target_path = save_path + \"/\" + ntpath.basename(trimd_file)\n",
    "    print(target_path)\n",
    "    shutil.copyfile(trimd_file, target_path)\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demultiplex Scifi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R21_trimPA.fastq.gz',\n",
       " '/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_RC_R21_trimPA.fastq.gz',\n",
       " '/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_R21_trimPA.fastq.gz',\n",
       " '/tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz',\n",
       " '/tmp/outdata/SciFi_old_cDNA_Pool2_R21_trimPA.fastq.gz']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#R21 used for demultiplexing - to get the well \n",
    "\n",
    "#scifi5 gfp pos i.e. DA cell wells, scifi 6, 7, 8 gfp neg\n",
    "\n",
    "'''\n",
    "nuclei taken from one animal and fac sorted. gfp pos nuclei put into one well, the other ones were put into different well and this \n",
    "was then split into 3 different wells to test different primer reactions. the primers found in index.fasta file\n",
    "RT index of the amplicon should correspond to the index array\n",
    "\n",
    "make sure your file does not contain the rev complements of the cell barcodes. the revcomp was used so that it can be used with the reagents\n",
    "in novaseq. \n",
    "\n",
    "solo\n",
    "\n",
    "'''\n",
    "\n",
    "samples_list = sorted(glob.glob(export_path+'*R21_trimPA.fastq.gz'))\n",
    "#samples_list= [item for item in samples_list if 'Undetermined' not in item]\n",
    "#samples_list= [item for item in samples_list if 'BC' in item]\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def deMultiFuse(sample_id) :\n",
    "    print(sample_id)\n",
    "    #sample_id = samples_list[6]\n",
    "    inPathR3 = export_path + ntpath.basename(sample_id).replace('_R21_trimPA.fastq.gz','_R3_trimPA.fastq.gz')\n",
    "    outPathR1 = export_path + ntpath.basename(sample_id).replace('_R21_trimPA.fastq.gz','_trimmed-{name}_R21.fastq.gz')\n",
    "    outPathR3 = export_path + ntpath.basename(sample_id).replace('_R21_trimPA.fastq.gz','_trimmed-{name}_R3.fastq.gz')\n",
    "    trimLog = export_path + ntpath.basename(sample_id).replace('_R21_trimPA.fastq.gz','_trimLog.tsv')\n",
    "    param = \"-e 0.15 --cores=0 --action none -a file:\\'\" + output_data2 + \"/Scifi6_index.fasta\\' --report=minimal\"\n",
    "    #param explanations: e, 0= max error rate and min. overlap for the adapters\n",
    "    call_sequence = \"cutadapt \"+param+\" -o \\'\"+outPathR1+\"\\' -p \\'\" + outPathR3 + \"\\' \\'\"+ sample_id + \"\\' \\'\" + inPathR3 +\"\\' > \\'\" + trimLog +\"\\'\"\n",
    "  #  print(call_sequence)\n",
    "    !($call_sequence)      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_Pool1-2_EnrichedBCs_All_R21_trimPA.fastq.gz\n",
      "[         8<-] 00:01:06    19,111,902 reads @   3.3 µs/read;  18.17 M reads/minute"
     ]
    }
   ],
   "source": [
    "[deMultiFuse(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate a QC table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>in_reads</th>\n",
       "      <th>in_bp</th>\n",
       "      <th>too_short</th>\n",
       "      <th>too_long</th>\n",
       "      <th>too_many_n</th>\n",
       "      <th>out_reads</th>\n",
       "      <th>w/adapters</th>\n",
       "      <th>qualtrim_bp</th>\n",
       "      <th>out_bp</th>\n",
       "      <th>w/adapters2</th>\n",
       "      <th>qualtrim2_bp</th>\n",
       "      <th>out2_bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SciFi_new_cDNA_Pool1</th>\n",
       "      <td>WARN</td>\n",
       "      <td>47181203</td>\n",
       "      <td>6950928856</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47181203</td>\n",
       "      <td>18326779</td>\n",
       "      <td>0</td>\n",
       "      <td>2000148204</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4950780652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SciFi_old_cDNA_Pool2</th>\n",
       "      <td>OK</td>\n",
       "      <td>19005056</td>\n",
       "      <td>2766519333</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19005056</td>\n",
       "      <td>17819921</td>\n",
       "      <td>0</td>\n",
       "      <td>816437301</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1950082032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     status  in_reads       in_bp  too_short  too_long  \\\n",
       "SciFi_new_cDNA_Pool1   WARN  47181203  6950928856          0         0   \n",
       "SciFi_old_cDNA_Pool2     OK  19005056  2766519333          0         0   \n",
       "\n",
       "                      too_many_n  out_reads  w/adapters  qualtrim_bp  \\\n",
       "SciFi_new_cDNA_Pool1           0   47181203    18326779            0   \n",
       "SciFi_old_cDNA_Pool2           0   19005056    17819921            0   \n",
       "\n",
       "                          out_bp  w/adapters2  qualtrim2_bp     out2_bp  \n",
       "SciFi_new_cDNA_Pool1  2000148204            0             0  4950780652  \n",
       "SciFi_old_cDNA_Pool2   816437301            0             0  1950082032  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_list = sorted(glob.glob(export_path+'*_trimLog.tsv'))\n",
    "allSampleTable = pd.DataFrame()\n",
    "\n",
    "for qc_file in qc_list :\n",
    "    df=pd.read_csv(qc_file, sep='\\t') # , header=0, index_col=0,skiprows=None\n",
    "    df = df.set_index(pd.Series(ntpath.basename(qc_file).replace('_trimLog.tsv','')))\n",
    "    allSampleTable = pd.concat([allSampleTable, df], axis=0)\n",
    "\n",
    "allSampleTable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove false cDNA indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi5_WP_R21.fastq.gz\n",
      "5379164\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi5_WP_R3.fastq.gz\n",
      "5379164\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi5_oDT_R21.fastq.gz\n",
      "46326\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi5_oDT_R3.fastq.gz\n",
      "46326\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi6_WP_R21.fastq.gz\n",
      "30490\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi6_WP_R3.fastq.gz\n",
      "30490\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi6_oDT_R21.fastq.gz\n",
      "163191\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi6_oDT_R3.fastq.gz\n",
      "163191\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi7_WP_R21.fastq.gz\n",
      "51268\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi7_WP_R3.fastq.gz\n",
      "51268\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi7_oDT_R21.fastq.gz\n",
      "27172\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi7_oDT_R3.fastq.gz\n",
      "27172\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi8_WP_R21.fastq.gz\n",
      "77589\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi8_WP_R3.fastq.gz\n",
      "77589\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi8_oDT_R21.fastq.gz\n",
      "2278810\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-SciFi8_oDT_R3.fastq.gz\n",
      "2278810\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP1_R21.fastq.gz\n",
      "40838\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP1_R3.fastq.gz\n",
      "40838\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP2_R21.fastq.gz\n",
      "7680896\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP2_R3.fastq.gz\n",
      "7680896\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP3_R21.fastq.gz\n",
      "41737\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP3_R3.fastq.gz\n",
      "41737\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP4_R21.fastq.gz\n",
      "161286\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-WP4_R3.fastq.gz\n",
      "161286\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT1_R21.fastq.gz\n",
      "2076425\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT1_R3.fastq.gz\n",
      "2076425\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT2_R21.fastq.gz\n",
      "19003\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT2_R3.fastq.gz\n",
      "19003\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT3_R21.fastq.gz\n",
      "217329\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT3_R3.fastq.gz\n",
      "217329\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT4_R21.fastq.gz\n",
      "35255\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-oDT4_R3.fastq.gz\n",
      "35255\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-unknown_R21.fastq.gz\n",
      "28854424\n",
      "/tmp/outdata/SciFi_new_cDNA_Pool1_trimmed-unknown_R3.fastq.gz\n",
      "28854424\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi5_WP_R21.fastq.gz\n",
      "144017\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi5_WP_R3.fastq.gz\n",
      "144017\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi5_oDT_R21.fastq.gz\n",
      "5502621\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi5_oDT_R3.fastq.gz\n",
      "5502621\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi6_WP_R21.fastq.gz\n",
      "809348\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi6_WP_R3.fastq.gz\n",
      "809348\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi6_oDT_R21.fastq.gz\n",
      "10963955\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi6_oDT_R3.fastq.gz\n",
      "10963955\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi7_WP_R21.fastq.gz\n",
      "40050\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi7_WP_R3.fastq.gz\n",
      "40050\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi7_oDT_R21.fastq.gz\n",
      "20570\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi7_oDT_R3.fastq.gz\n",
      "20570\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi8_WP_R21.fastq.gz\n",
      "29309\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi8_WP_R3.fastq.gz\n",
      "29309\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi8_oDT_R21.fastq.gz\n",
      "35630\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-SciFi8_oDT_R3.fastq.gz\n",
      "35630\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP1_R21.fastq.gz\n",
      "41041\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP1_R3.fastq.gz\n",
      "41041\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP2_R21.fastq.gz\n",
      "19310\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP2_R3.fastq.gz\n",
      "19310\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP3_R21.fastq.gz\n",
      "30264\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP3_R3.fastq.gz\n",
      "30264\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP4_R21.fastq.gz\n",
      "95637\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-WP4_R3.fastq.gz\n",
      "95637\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT1_R21.fastq.gz\n",
      "50708\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT1_R3.fastq.gz\n",
      "50708\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT2_R21.fastq.gz\n",
      "27173\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT2_R3.fastq.gz\n",
      "27173\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT3_R21.fastq.gz\n",
      "6259\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT3_R3.fastq.gz\n",
      "6259\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT4_R21.fastq.gz\n",
      "4029\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-oDT4_R3.fastq.gz\n",
      "4029\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-unknown_R21.fastq.gz\n",
      "1185135\n",
      "/tmp/outdata/SciFi_old_cDNA_Pool2_trimmed-unknown_R3.fastq.gz\n",
      "1185135\n"
     ]
    }
   ],
   "source": [
    "#removal of sequencing artifacts\n",
    "samples_list = sorted(glob.glob(export_path+'*trimmed*.fastq.gz'))\n",
    "for sample_file in samples_list :\n",
    "    print(sample_file)\n",
    "    result = subprocess.run(['pigz -dc '+ sample_file+ ' | wc -l'], stdout=subprocess.PIPE, shell=True)\n",
    "    sampleReads = int(int(result.stdout)/4)\n",
    "    if sampleReads < 4000 :\n",
    "        os.remove(sample_file)\n",
    "    print(sampleReads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List selected samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi5_WP_R21.fastq.gz\n",
      "11607682\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi5_oDT_R21.fastq.gz\n",
      "195116926\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi6_WP_R21.fastq.gz\n",
      "3590030\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi6_oDT_R21.fastq.gz\n",
      "48129839\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi7_WP_R21.fastq.gz\n",
      "5420628\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi7_oDT_R21.fastq.gz\n",
      "19880142\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi8_WP_R21.fastq.gz\n",
      "4423672\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-SciFi8_oDT_R21.fastq.gz\n",
      "17860185\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-WP1_R21.fastq.gz\n",
      "452255\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-WP2_R21.fastq.gz\n",
      "357452\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-WP3_R21.fastq.gz\n",
      "464263\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-WP4_R21.fastq.gz\n",
      "1557606\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-oDT1_R21.fastq.gz\n",
      "494403\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-oDT2_R21.fastq.gz\n",
      "197908\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-oDT3_R21.fastq.gz\n",
      "170709\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-oDT4_R21.fastq.gz\n",
      "137331\n",
      "/tmp/outdata/Scifi_library_2_S2_trimmed-unknown_R21.fastq.gz\n",
      "60285405\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi5_WP_R21.fastq.gz\n",
      "2796212\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi5_oDT_R21.fastq.gz\n",
      "9869193\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi6_WP_R21.fastq.gz\n",
      "243556\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi6_oDT_R21.fastq.gz\n",
      "2485225\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi7_WP_R21.fastq.gz\n",
      "469891\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi7_oDT_R21.fastq.gz\n",
      "883390\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi8_WP_R21.fastq.gz\n",
      "397314\n",
      "/tmp/outdata/Undetermined_S0_trimmed-SciFi8_oDT_R21.fastq.gz\n",
      "1236577\n",
      "/tmp/outdata/Undetermined_S0_trimmed-WP1_R21.fastq.gz\n",
      "380558\n",
      "/tmp/outdata/Undetermined_S0_trimmed-WP2_R21.fastq.gz\n",
      "122145\n",
      "/tmp/outdata/Undetermined_S0_trimmed-WP3_R21.fastq.gz\n",
      "328424\n",
      "/tmp/outdata/Undetermined_S0_trimmed-WP4_R21.fastq.gz\n",
      "220366\n",
      "/tmp/outdata/Undetermined_S0_trimmed-oDT1_R21.fastq.gz\n",
      "406771\n",
      "/tmp/outdata/Undetermined_S0_trimmed-oDT2_R21.fastq.gz\n",
      "136100\n",
      "/tmp/outdata/Undetermined_S0_trimmed-oDT3_R21.fastq.gz\n",
      "183662\n",
      "/tmp/outdata/Undetermined_S0_trimmed-oDT4_R21.fastq.gz\n",
      "35300\n",
      "/tmp/outdata/Undetermined_S0_trimmed-unknown_R21.fastq.gz\n",
      "9922464\n"
     ]
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(export_path+'*trimmed*R21.fastq.gz'))\n",
    "for sample_file in samples_list :\n",
    "    print(sample_file)\n",
    "    #pigz -d decompressed the sample file and then we provide the number of lines \n",
    "    result = subprocess.run(['pigz -dc '+ sample_file+ ' | wc -l'], stdout=subprocess.PIPE, shell=True)\n",
    "    sampleReads = int(int(result.stdout)/4)\n",
    "    print(sampleReads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge all individual files from the same cDNA index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#by same cdna index does it refer to e.g oDT?\n",
    "\n",
    "#not needed for scifi6 as already done prior!\n",
    "'''\n",
    "def catSameID(cDNA_id) :\n",
    "    print(cDNA_id)\n",
    "    catSeq = \"cat \"+export_path+\"*oDT_\"+cDNA_id+\"_R21.fastq.gz > /tmp/outdata/SciFi_oDT_\" +cDNA_id+\"_R21.fastq.gz\"\n",
    "    catSeqRev = \"cat \"+export_path+\"*oDT_\"+cDNA_id+\"_R3.fastq.gz > /tmp/outdata/SciFi_oDT_\" +cDNA_id+\"_R3.fastq.gz\"\n",
    "    print(catSeq)\n",
    "    !($catSeq)\n",
    "    !($catSeqRev) \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1_5\n",
      "cat /tmp/outdata/*oDT_A1_5_R21.fastq.gz > /tmp/outdata/SciFi_oDT_A1_5_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_A1_5_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_A1_5_R3.fastq.gz': No such file or directory\n",
      "B1_57\n",
      "cat /tmp/outdata/*oDT_B1_57_R21.fastq.gz > /tmp/outdata/SciFi_oDT_B1_57_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_B1_57_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_B1_57_R3.fastq.gz': No such file or directory\n",
      "C1_97\n",
      "cat /tmp/outdata/*oDT_C1_97_R21.fastq.gz > /tmp/outdata/SciFi_oDT_C1_97_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_C1_97_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_C1_97_R3.fastq.gz': No such file or directory\n",
      "D1_137\n",
      "cat /tmp/outdata/*oDT_D1_137_R21.fastq.gz > /tmp/outdata/SciFi_oDT_D1_137_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_D1_137_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_D1_137_R3.fastq.gz': No such file or directory\n",
      "E1_182\n",
      "cat /tmp/outdata/*oDT_E1_182_R21.fastq.gz > /tmp/outdata/SciFi_oDT_E1_182_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_E1_182_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_E1_182_R3.fastq.gz': No such file or directory\n",
      "F1_227\n",
      "cat /tmp/outdata/*oDT_F1_227_R21.fastq.gz > /tmp/outdata/SciFi_oDT_F1_227_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_F1_227_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_F1_227_R3.fastq.gz': No such file or directory\n",
      "G1_266\n",
      "cat /tmp/outdata/*oDT_G1_266_R21.fastq.gz > /tmp/outdata/SciFi_oDT_G1_266_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_G1_266_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_G1_266_R3.fastq.gz': No such file or directory\n",
      "G2_271\n",
      "cat /tmp/outdata/*oDT_G2_271_R21.fastq.gz > /tmp/outdata/SciFi_oDT_G2_271_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_G2_271_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_G2_271_R3.fastq.gz': No such file or directory\n",
      "H1_307\n",
      "cat /tmp/outdata/*oDT_H1_307_R21.fastq.gz > /tmp/outdata/SciFi_oDT_H1_307_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_H1_307_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_H1_307_R3.fastq.gz': No such file or directory\n",
      "H2_314\n",
      "cat /tmp/outdata/*oDT_H2_314_R21.fastq.gz > /tmp/outdata/SciFi_oDT_H2_314_R21.fastq.gz\n",
      "cat: '/tmp/outdata/*oDT_H2_314_R21.fastq.gz': No such file or directory\n",
      "cat: '/tmp/outdata/*oDT_H2_314_R3.fastq.gz': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "#Where do these ids come from?\n",
    "catSameID(\"A1_5\")\n",
    "catSameID(\"B1_57\")\n",
    "catSameID(\"C1_97\")\n",
    "catSameID(\"D1_137\")\n",
    "catSameID(\"E1_182\")\n",
    "catSameID(\"F1_227\")\n",
    "catSameID(\"G1_266\")\n",
    "catSameID(\"G2_271\")\n",
    "catSameID(\"H1_307\")\n",
    "catSameID(\"H2_314\")\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat /tmp/outdata/*R21_trimPA.fastq.gz > /tmp/outdata/SciFi_oDT_Complete_R21_trimPA.fastq.gz\n",
    "!cat /tmp/outdata/*R3_trimPA.fastq.gz > /tmp/outdata/SciFi_oDT_Complete_R3_trimPA.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install STAR and align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## The released versions up and including 2.7.9a contain a bug in the STARsolo for quantification. This it needs to be built from pulled from Github and built from source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "make -C htslib lib-static\n",
      "make[1]: Entering directory '/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source/htslib'\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o kfunc.o kfunc.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o knetfile.o knetfile.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o kstring.o kstring.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o bgzf.o bgzf.c\n",
      "\u001b[01m\u001b[Kbgzf.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kbgzf_write\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kbgzf.c:733:5:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kthis ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’ clause does not guard... [\u001b[01;35m\u001b[K-Wmisleading-indentation\u001b[m\u001b[K]\n",
      "  733 |     \u001b[01;35m\u001b[Kif\u001b[m\u001b[K ( !fp->is_compressed )\n",
      "      |     \u001b[01;35m\u001b[K^~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kbgzf.c:736:2:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[K...this statement, but the latter is misleadingly indented as if it were guarded by the ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’\n",
      "  736 |  \u001b[01;36m\u001b[Kconst\u001b[m\u001b[K uint8_t *input = (const uint8_t*)data;\n",
      "      |  \u001b[01;36m\u001b[K^~~~~\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o faidx.o faidx.c\n",
      "\u001b[01m\u001b[Kfaidx.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kfai_load\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kfaidx.c:265:5:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kthis ‘\u001b[01m\u001b[Kelse\u001b[m\u001b[K’ clause does not guard... [\u001b[01;35m\u001b[K-Wmisleading-indentation\u001b[m\u001b[K]\n",
      "  265 |     \u001b[01;35m\u001b[Kelse\u001b[m\u001b[K\n",
      "      |     \u001b[01;35m\u001b[K^~~~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kfaidx.c:268:2:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[K...this statement, but the latter is misleadingly indented as if it were guarded by the ‘\u001b[01m\u001b[Kelse\u001b[m\u001b[K’\n",
      "  268 |  \u001b[01;36m\u001b[Kif\u001b[m\u001b[K (fp == 0) {\n",
      "      |  \u001b[01;36m\u001b[K^~\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o hfile.o hfile.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o hfile_net.o hfile_net.c\n",
      "echo '#define HTS_VERSION \"0.0.1\"' > version.h\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o hts.o hts.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o sam.o sam.c\n",
      "In file included from \u001b[01m\u001b[K/usr/include/string.h:495\u001b[m\u001b[K,\n",
      "                 from \u001b[01m\u001b[Ksam.c:3\u001b[m\u001b[K:\n",
      "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
      "    inlined from ‘\u001b[01m\u001b[Kbam_hdr_write\u001b[m\u001b[K’ at \u001b[01m\u001b[Ksam.c:130:2\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ output truncated before terminating nul copying 4 bytes from a string of the same length [\u001b[01;35m\u001b[K-Wstringop-truncation\u001b[m\u001b[K]\n",
      "  106 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len, __bos (__dest))\u001b[m\u001b[K;\n",
      "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o vcf_sweep.o vcf_sweep.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o tbx.o tbx.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o vcf.o vcf.c\n",
      "\u001b[01m\u001b[Kvcf.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kvcf_format\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kvcf.c:1909:4:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kthis ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’ clause does not guard... [\u001b[01;35m\u001b[K-Wmisleading-indentation\u001b[m\u001b[K]\n",
      " 1909 |    \u001b[01;35m\u001b[Kif\u001b[m\u001b[K ( !first ) kputc(';', s); first = 0;\n",
      "      |    \u001b[01;35m\u001b[K^~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kvcf.c:1909:33:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[K...this statement, but the latter is misleadingly indented as if it were guarded by the ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’\n",
      " 1909 |    if ( !first ) kputc(';', s); \u001b[01;36m\u001b[Kfirst\u001b[m\u001b[K = 0;\n",
      "      |                                 \u001b[01;36m\u001b[K^~~~~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kvcf.c:1948:21:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kthis ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’ clause does not guard... [\u001b[01;35m\u001b[K-Wmisleading-indentation\u001b[m\u001b[K]\n",
      " 1948 |                     \u001b[01;35m\u001b[Kif\u001b[m\u001b[K (!first) kputc(':', s); first = 0;\n",
      "      |                     \u001b[01;35m\u001b[K^~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kvcf.c:1948:48:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[K...this statement, but the latter is misleadingly indented as if it were guarded by the ‘\u001b[01m\u001b[Kif\u001b[m\u001b[K’\n",
      " 1948 |                     if (!first) kputc(':', s); \u001b[01;36m\u001b[Kfirst\u001b[m\u001b[K = 0;\n",
      "      |                                                \u001b[01;36m\u001b[K^~~~~\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o vcfutils.o vcfutils.c\n",
      "In file included from \u001b[01m\u001b[Khtslib/vcf.h:12\u001b[m\u001b[K,\n",
      "                 from \u001b[01m\u001b[Khtslib/vcfutils.h:8\u001b[m\u001b[K,\n",
      "                 from \u001b[01m\u001b[Kvcfutils.c:1\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[Kvcfutils.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kbcf_remove_alleles\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kvcfutils.c:291:25:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Ksuggest parentheses around assignment used as truth value [\u001b[01;35m\u001b[K-Wparentheses\u001b[m\u001b[K]\n",
      "  291 |                 assert( \u001b[01;35m\u001b[Kn\u001b[m\u001b[K=nG_ori );\n",
      "      |                         \u001b[01;35m\u001b[K^\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_decode.o cram/cram_decode.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_encode.o cram/cram_encode.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_index.o cram/cram_index.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_io.o cram/cram_io.c\n",
      "\u001b[01m\u001b[Kcram/cram_io.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kcram_populate_ref\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kcram/cram_io.c:1529:27:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K.tmp_\u001b[m\u001b[K’ directive writing 5 bytes into a region of size between 1 and 4096 [\u001b[01;35m\u001b[K-Wformat-overflow=\u001b[m\u001b[K]\n",
      " 1529 |      sprintf(path_tmp, \"%s\u001b[01;35m\u001b[K.tmp_\u001b[m\u001b[K%d\", path, /*getpid(),*/ i);\n",
      "      |                           \u001b[01;35m\u001b[K^~~~~\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kcram/cram_io.c:1529:24:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[Kdirective argument in the range [0, 2147483647]\n",
      " 1529 |      sprintf(path_tmp, \u001b[01;36m\u001b[K\"%s.tmp_%d\"\u001b[m\u001b[K, path, /*getpid(),*/ i);\n",
      "      |                        \u001b[01;36m\u001b[K^~~~~~~~~~~\u001b[m\u001b[K\n",
      "In file included from \u001b[01m\u001b[K/usr/include/stdio.h:867\u001b[m\u001b[K,\n",
      "                 from \u001b[01m\u001b[Kcram/cram_io.c:51\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/stdio2.h:36:10:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin___sprintf_chk\u001b[m\u001b[K’ output between 7 and 4111 bytes into a destination of size 4096\n",
      "   36 |   return \u001b[01;36m\u001b[K__builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,\u001b[m\u001b[K\n",
      "      |          \u001b[01;36m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "   37 | \u001b[01;36m\u001b[K      __bos (__s), __fmt, __va_arg_pack ())\u001b[m\u001b[K;\n",
      "      |       \u001b[01;36m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "In file included from \u001b[01m\u001b[K/usr/include/string.h:495\u001b[m\u001b[K,\n",
      "                 from \u001b[01m\u001b[Kcram/cram_io.c:55\u001b[m\u001b[K:\n",
      "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
      "    inlined from ‘\u001b[01m\u001b[Kfull_path\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:2917:6\u001b[m\u001b[K,\n",
      "    inlined from ‘\u001b[01m\u001b[Kcram_write_SAM_hdr\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:2987:3\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 4096 equals destination size [\u001b[01;35m\u001b[K-Wstringop-truncation\u001b[m\u001b[K]\n",
      "  106 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len, __bos (__dest))\u001b[m\u001b[K;\n",
      "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
      "    inlined from ‘\u001b[01m\u001b[Kfull_path\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:2909:2\u001b[m\u001b[K,\n",
      "    inlined from ‘\u001b[01m\u001b[Kcram_write_SAM_hdr\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:2987:3\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 4096 equals destination size [\u001b[01;35m\u001b[K-Wstringop-truncation\u001b[m\u001b[K]\n",
      "  106 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len, __bos (__dest))\u001b[m\u001b[K;\n",
      "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
      "    inlined from ‘\u001b[01m\u001b[Kcram_dopen\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:3291:2\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 20 equals destination size [\u001b[01;35m\u001b[K-Wstringop-truncation\u001b[m\u001b[K]\n",
      "  106 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len, __bos (__dest))\u001b[m\u001b[K;\n",
      "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_samtools.o cram/cram_samtools.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/cram_stats.o cram/cram_stats.c\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/files.o cram/files.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/mFILE.o cram/mFILE.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/md5.o cram/md5.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/sam_header.o cram/sam_header.c\n",
      "In file included from \u001b[01m\u001b[Kcram/sam_header.c:36\u001b[m\u001b[K:\n",
      "\u001b[01m\u001b[Kcram/sam_header.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Ksam_hdr_add_lines\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kcram/sam_header.c:328:13:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Ksuggest parentheses around assignment used as truth value [\u001b[01;35m\u001b[K-Wparentheses\u001b[m\u001b[K]\n",
      "  328 |      assert(\u001b[01;35m\u001b[Kp\u001b[m\u001b[K->next = t);\n",
      "      |             \u001b[01;35m\u001b[K^\u001b[m\u001b[K\n",
      "\u001b[01m\u001b[Kcram/sam_header.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Ksam_hdr_vadd\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kcram/sam_header.c:449:9:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Ksuggest parentheses around assignment used as truth value [\u001b[01;35m\u001b[K-Wparentheses\u001b[m\u001b[K]\n",
      "  449 |  assert(\u001b[01;35m\u001b[Kp\u001b[m\u001b[K->next = t);\n",
      "      |         \u001b[01;35m\u001b[K^\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/string_alloc.o cram/string_alloc.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/thread_pool.o cram/thread_pool.c\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/vlen.o cram/vlen.c\n",
      "\u001b[01m\u001b[Kcram/vlen.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kvflen\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[Kcram/vlen.c:121:9:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[Ki\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K-Wunused-but-set-variable\u001b[m\u001b[K]\n",
      "  121 |     int \u001b[01;35m\u001b[Ki\u001b[m\u001b[K;\n",
      "      |         \u001b[01;35m\u001b[K^\u001b[m\u001b[K\n",
      "cc -g -Wall -O2  -I. -DSAMTOOLS=1 -c -o cram/zfio.o cram/zfio.c\n",
      "ar -rc libhts.a kfunc.o knetfile.o kstring.o bgzf.o faidx.o hfile.o hfile_net.o hts.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_index.o cram/cram_io.o cram/cram_samtools.o cram/cram_stats.o cram/files.o cram/mFILE.o cram/md5.o cram/open_trace_file.o cram/pooled_alloc.o cram/sam_header.o cram/string_alloc.o cram/thread_pool.o cram/vlen.o cram/zfio.o\n",
      "ranlib libhts.a\n",
      "make[1]: Leaving directory '/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source/htslib'\n",
      "echo ReadAlign_mapOneRead.cpp readLoad.cpp ReadAlign_mapOneReadSpliceGraph.cpp SoloReadFeature.cpp ChimericDetection_chimericDetectionMult.cpp sjdbPrepare.cpp ChimericSegment.cpp Transcriptome_classifyAlign.cpp ReadAlign_stitchPieces.cpp funCompareUintAndSuffixes.cpp SoloFeature_sumThreads.cpp SoloFeature.cpp SoloBarcode_extractBarcode.cpp Parameters_readSAMheader.cpp BAMbinSortByCoordinate.cpp Transcriptome_quantAlign.cpp ReadAlign_alignBAM.cpp GTF_transcriptGeneSJ.cpp SoloFeature_processRecords.cpp SpliceGraph_swTraceBack.cpp Transcriptome.cpp Chain.cpp SoloFeature_emptyDrops_CR.cpp Stats.cpp Transcript_generateCigarP.cpp SoloReadFeature_record.cpp Transcript_transformGenome.cpp SoloFeature_countVelocyto.cpp Parameters_openReadsFiles.cpp ReadAlign_outputSpliceGraphSAM.cpp ReadAlignChunk.cpp stitchAlignToTranscript.cpp ChimericAlign_chimericStitching.cpp signalFromBAM.cpp sjAlignSplit.cpp SoloFeature_statsOutput.cpp Genome_genomeLoad.cpp stitchWindowAligns.cpp ParametersChimeric_initialize.cpp BAMfunctions.cpp ReadAlign_maxMappableLength2strands.cpp SoloFeature_countSmartSeq.cpp soloInputFeatureUMI.cpp Genome_insertSequences.cpp ReadAlign_chimericDetectionPEmerged.cpp insertSeqSA.cpp streamFuns.cpp ReadAlignChunk_processChunks.cpp BAMoutput.cpp SpliceGraph_findSuperTr.cpp Transcript_alignScore.cpp GTF.cpp SoloReadBarcode.cpp ReadAlign_CIGAR.cpp ParametersSolo.cpp Parameters_readFilesInit.cpp ReadAlign_outputVariation.cpp funCompareUintAndSuffixesMemcmp.cpp samHeaders.cpp ReadAlign_chimericDetectionOldOutput.cpp mapThreadsSpawn.cpp SuperTranscriptome.cpp stringSubstituteAll.cpp Parameters_samAttributes.cpp SharedMemory.cpp ReadAlign_waspMap.cpp Genome_genomeGenerate.cpp ReadAlign_storeAligns.cpp ParametersClip_initialize.cpp ReadAlign_stitchWindowSeeds.cpp ClipMate_initialize.cpp SuffixArrayFuns.cpp SequenceFuns.cpp extendAlign.cpp ReadAlign_transformGenome.cpp sjdbLoadFromFiles.cpp Transcriptome_alignExonOverlap.cpp genomeScanFastaFiles.cpp sjdbBuildIndex.cpp Genome.cpp Variation.cpp ReadAlign_outputTranscriptSAM.cpp ChimericAlign_chimericJunctionOutput.cpp binarySearch2.cpp sjdbInsertJunctions.cpp Transcript_variationAdjust.cpp ReadAlign_peOverlapMergeMap.cpp Parameters.cpp SoloFeature_collapseUMI_Graph.cpp SoloReadFeature_inputRecords.cpp ClipMate_clip.cpp STAR.cpp bamSortByCoordinate.cpp Solo.cpp ReadAlign_outputAlignments.cpp SpliceGraph.cpp SoloFeature_outputResults.cpp SoloRead.cpp SoloFeature_loadRawMatrix.cpp OutSJ.cpp ReadAlign_outputTranscriptCIGARp.cpp ErrorWarning.cpp SoloBarcode.cpp ChimericAlign.cpp ChimericAlign_chimericBAMoutput.cpp serviceFuns.cpp TimeFunctions.cpp stitchGapIndel.cpp genomeSAindex.cpp ReadAlign_calcCIGAR.cpp sjdbLoadFromStream.cpp Parameters_closeReadsFiles.cpp Genome_genomeOutLoad.cpp ParametersGenome.cpp Transcript_convertGenomeCigar.cpp SoloRead_record.cpp outputSJ.cpp SoloFeature_countCBgeneUMI.cpp ReadAlignChunk_mapChunk.cpp ThreadControl.cpp SoloFeature_addBAMtags.cpp ReadAlign_createExtendWindowsWithAlign.cpp genomeParametersWrite.cpp ReadAlign_multMapSelect.cpp ClipMate_clipChunk.cpp ReadAlign_chimericDetectionOld.cpp ReadAlign_chimericDetection.cpp ReadAlign_outputTranscriptSJ.cpp SoloFeature_redistributeReadsByCB.cpp ReadAlign.cpp Transcriptome_geneFullAlignOverlap.cpp ReadAlign_assignAlignToWindow.cpp sysRemoveDir.cpp BAMbinSortUnmapped.cpp ChimericDetection.cpp GlobalVariables.cpp twoPassRunPass1.cpp Transcriptome_geneCountsAddAlign.cpp ReadAlign_mappedFilter.cpp SpliceGraph_swScoreSpliced.cpp Quantifications.cpp PackedArray.cpp bamRemoveDuplicates.cpp InOutStreams.cpp ReadAlign_oneRead.cpp Transcriptome_geneFullAlignOverlap_ExonOverIntron.cpp SoloReadBarcode_getCBandUMI.cpp Transcript_variationOutput.cpp Transcript.cpp GTF_superTranscript.cpp SoloFeature_collapseUMIall_multiMappers.cpp blocksOverlap.cpp ReadAlign_quantTranscriptome.cpp SoloFeature_cellFiltering.cpp SoloFeature_quantTranscript.cpp ClipCR4.cpp Genome_transformGenome.cpp bam_cat.c\n",
      "ReadAlign_mapOneRead.cpp readLoad.cpp ReadAlign_mapOneReadSpliceGraph.cpp SoloReadFeature.cpp ChimericDetection_chimericDetectionMult.cpp sjdbPrepare.cpp ChimericSegment.cpp Transcriptome_classifyAlign.cpp ReadAlign_stitchPieces.cpp funCompareUintAndSuffixes.cpp SoloFeature_sumThreads.cpp SoloFeature.cpp SoloBarcode_extractBarcode.cpp Parameters_readSAMheader.cpp BAMbinSortByCoordinate.cpp Transcriptome_quantAlign.cpp ReadAlign_alignBAM.cpp GTF_transcriptGeneSJ.cpp SoloFeature_processRecords.cpp SpliceGraph_swTraceBack.cpp Transcriptome.cpp Chain.cpp SoloFeature_emptyDrops_CR.cpp Stats.cpp Transcript_generateCigarP.cpp SoloReadFeature_record.cpp Transcript_transformGenome.cpp SoloFeature_countVelocyto.cpp Parameters_openReadsFiles.cpp ReadAlign_outputSpliceGraphSAM.cpp ReadAlignChunk.cpp stitchAlignToTranscript.cpp ChimericAlign_chimericStitching.cpp signalFromBAM.cpp sjAlignSplit.cpp SoloFeature_statsOutput.cpp Genome_genomeLoad.cpp stitchWindowAligns.cpp ParametersChimeric_initialize.cpp BAMfunctions.cpp ReadAlign_maxMappableLength2strands.cpp SoloFeature_countSmartSeq.cpp soloInputFeatureUMI.cpp Genome_insertSequences.cpp ReadAlign_chimericDetectionPEmerged.cpp insertSeqSA.cpp streamFuns.cpp ReadAlignChunk_processChunks.cpp BAMoutput.cpp SpliceGraph_findSuperTr.cpp Transcript_alignScore.cpp GTF.cpp SoloReadBarcode.cpp ReadAlign_CIGAR.cpp ParametersSolo.cpp Parameters_readFilesInit.cpp ReadAlign_outputVariation.cpp funCompareUintAndSuffixesMemcmp.cpp samHeaders.cpp ReadAlign_chimericDetectionOldOutput.cpp mapThreadsSpawn.cpp SuperTranscriptome.cpp stringSubstituteAll.cpp Parameters_samAttributes.cpp SharedMemory.cpp ReadAlign_waspMap.cpp Genome_genomeGenerate.cpp ReadAlign_storeAligns.cpp ParametersClip_initialize.cpp ReadAlign_stitchWindowSeeds.cpp ClipMate_initialize.cpp SuffixArrayFuns.cpp SequenceFuns.cpp extendAlign.cpp ReadAlign_transformGenome.cpp sjdbLoadFromFiles.cpp Transcriptome_alignExonOverlap.cpp genomeScanFastaFiles.cpp sjdbBuildIndex.cpp Genome.cpp Variation.cpp ReadAlign_outputTranscriptSAM.cpp ChimericAlign_chimericJunctionOutput.cpp binarySearch2.cpp sjdbInsertJunctions.cpp Transcript_variationAdjust.cpp ReadAlign_peOverlapMergeMap.cpp Parameters.cpp SoloFeature_collapseUMI_Graph.cpp SoloReadFeature_inputRecords.cpp ClipMate_clip.cpp STAR.cpp bamSortByCoordinate.cpp Solo.cpp ReadAlign_outputAlignments.cpp SpliceGraph.cpp SoloFeature_outputResults.cpp SoloRead.cpp SoloFeature_loadRawMatrix.cpp OutSJ.cpp ReadAlign_outputTranscriptCIGARp.cpp ErrorWarning.cpp SoloBarcode.cpp ChimericAlign.cpp ChimericAlign_chimericBAMoutput.cpp serviceFuns.cpp TimeFunctions.cpp stitchGapIndel.cpp genomeSAindex.cpp ReadAlign_calcCIGAR.cpp sjdbLoadFromStream.cpp Parameters_closeReadsFiles.cpp Genome_genomeOutLoad.cpp ParametersGenome.cpp Transcript_convertGenomeCigar.cpp SoloRead_record.cpp outputSJ.cpp SoloFeature_countCBgeneUMI.cpp ReadAlignChunk_mapChunk.cpp ThreadControl.cpp SoloFeature_addBAMtags.cpp ReadAlign_createExtendWindowsWithAlign.cpp genomeParametersWrite.cpp ReadAlign_multMapSelect.cpp ClipMate_clipChunk.cpp ReadAlign_chimericDetectionOld.cpp ReadAlign_chimericDetection.cpp ReadAlign_outputTranscriptSJ.cpp SoloFeature_redistributeReadsByCB.cpp ReadAlign.cpp Transcriptome_geneFullAlignOverlap.cpp ReadAlign_assignAlignToWindow.cpp sysRemoveDir.cpp BAMbinSortUnmapped.cpp ChimericDetection.cpp GlobalVariables.cpp twoPassRunPass1.cpp Transcriptome_geneCountsAddAlign.cpp ReadAlign_mappedFilter.cpp SpliceGraph_swScoreSpliced.cpp Quantifications.cpp PackedArray.cpp bamRemoveDuplicates.cpp InOutStreams.cpp ReadAlign_oneRead.cpp Transcriptome_geneFullAlignOverlap_ExonOverIntron.cpp SoloReadBarcode_getCBandUMI.cpp Transcript_variationOutput.cpp Transcript.cpp GTF_superTranscript.cpp SoloFeature_collapseUMIall_multiMappers.cpp blocksOverlap.cpp ReadAlign_quantTranscriptome.cpp SoloFeature_cellFiltering.cpp SoloFeature_quantTranscript.cpp ClipCR4.cpp Genome_transformGenome.cpp bam_cat.c\n",
      "'rm' -f ./Depend.list\n",
      "g++ -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:10+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -MM ReadAlign_mapOneRead.cpp readLoad.cpp ReadAlign_mapOneReadSpliceGraph.cpp SoloReadFeature.cpp ChimericDetection_chimericDetectionMult.cpp sjdbPrepare.cpp ChimericSegment.cpp Transcriptome_classifyAlign.cpp ReadAlign_stitchPieces.cpp funCompareUintAndSuffixes.cpp SoloFeature_sumThreads.cpp SoloFeature.cpp SoloBarcode_extractBarcode.cpp Parameters_readSAMheader.cpp BAMbinSortByCoordinate.cpp Transcriptome_quantAlign.cpp ReadAlign_alignBAM.cpp GTF_transcriptGeneSJ.cpp SoloFeature_processRecords.cpp SpliceGraph_swTraceBack.cpp Transcriptome.cpp Chain.cpp SoloFeature_emptyDrops_CR.cpp Stats.cpp Transcript_generateCigarP.cpp SoloReadFeature_record.cpp Transcript_transformGenome.cpp SoloFeature_countVelocyto.cpp Parameters_openReadsFiles.cpp ReadAlign_outputSpliceGraphSAM.cpp ReadAlignChunk.cpp stitchAlignToTranscript.cpp ChimericAlign_chimericStitching.cpp signalFromBAM.cpp sjAlignSplit.cpp SoloFeature_statsOutput.cpp Genome_genomeLoad.cpp stitchWindowAligns.cpp ParametersChimeric_initialize.cpp BAMfunctions.cpp ReadAlign_maxMappableLength2strands.cpp SoloFeature_countSmartSeq.cpp soloInputFeatureUMI.cpp Genome_insertSequences.cpp ReadAlign_chimericDetectionPEmerged.cpp insertSeqSA.cpp streamFuns.cpp ReadAlignChunk_processChunks.cpp BAMoutput.cpp SpliceGraph_findSuperTr.cpp Transcript_alignScore.cpp GTF.cpp SoloReadBarcode.cpp ReadAlign_CIGAR.cpp ParametersSolo.cpp Parameters_readFilesInit.cpp ReadAlign_outputVariation.cpp funCompareUintAndSuffixesMemcmp.cpp samHeaders.cpp ReadAlign_chimericDetectionOldOutput.cpp mapThreadsSpawn.cpp SuperTranscriptome.cpp stringSubstituteAll.cpp Parameters_samAttributes.cpp SharedMemory.cpp ReadAlign_waspMap.cpp Genome_genomeGenerate.cpp ReadAlign_storeAligns.cpp ParametersClip_initialize.cpp ReadAlign_stitchWindowSeeds.cpp ClipMate_initialize.cpp SuffixArrayFuns.cpp SequenceFuns.cpp extendAlign.cpp ReadAlign_transformGenome.cpp sjdbLoadFromFiles.cpp Transcriptome_alignExonOverlap.cpp genomeScanFastaFiles.cpp sjdbBuildIndex.cpp Genome.cpp Variation.cpp ReadAlign_outputTranscriptSAM.cpp ChimericAlign_chimericJunctionOutput.cpp binarySearch2.cpp sjdbInsertJunctions.cpp Transcript_variationAdjust.cpp ReadAlign_peOverlapMergeMap.cpp Parameters.cpp SoloFeature_collapseUMI_Graph.cpp SoloReadFeature_inputRecords.cpp ClipMate_clip.cpp STAR.cpp bamSortByCoordinate.cpp Solo.cpp ReadAlign_outputAlignments.cpp SpliceGraph.cpp SoloFeature_outputResults.cpp SoloRead.cpp SoloFeature_loadRawMatrix.cpp OutSJ.cpp ReadAlign_outputTranscriptCIGARp.cpp ErrorWarning.cpp SoloBarcode.cpp ChimericAlign.cpp ChimericAlign_chimericBAMoutput.cpp serviceFuns.cpp TimeFunctions.cpp stitchGapIndel.cpp genomeSAindex.cpp ReadAlign_calcCIGAR.cpp sjdbLoadFromStream.cpp Parameters_closeReadsFiles.cpp Genome_genomeOutLoad.cpp ParametersGenome.cpp Transcript_convertGenomeCigar.cpp SoloRead_record.cpp outputSJ.cpp SoloFeature_countCBgeneUMI.cpp ReadAlignChunk_mapChunk.cpp ThreadControl.cpp SoloFeature_addBAMtags.cpp ReadAlign_createExtendWindowsWithAlign.cpp genomeParametersWrite.cpp ReadAlign_multMapSelect.cpp ClipMate_clipChunk.cpp ReadAlign_chimericDetectionOld.cpp ReadAlign_chimericDetection.cpp ReadAlign_outputTranscriptSJ.cpp SoloFeature_redistributeReadsByCB.cpp ReadAlign.cpp Transcriptome_geneFullAlignOverlap.cpp ReadAlign_assignAlignToWindow.cpp sysRemoveDir.cpp BAMbinSortUnmapped.cpp ChimericDetection.cpp GlobalVariables.cpp twoPassRunPass1.cpp Transcriptome_geneCountsAddAlign.cpp ReadAlign_mappedFilter.cpp SpliceGraph_swScoreSpliced.cpp Quantifications.cpp PackedArray.cpp bamRemoveDuplicates.cpp InOutStreams.cpp ReadAlign_oneRead.cpp Transcriptome_geneFullAlignOverlap_ExonOverIntron.cpp SoloReadBarcode_getCBandUMI.cpp Transcript_variationOutput.cpp Transcript.cpp GTF_superTranscript.cpp SoloFeature_collapseUMIall_multiMappers.cpp blocksOverlap.cpp ReadAlign_quantTranscriptome.cpp SoloFeature_cellFiltering.cpp SoloFeature_quantTranscript.cpp ClipCR4.cpp Genome_transformGenome.cpp bam_cat.c parametersDefault.xxd htslib >> Depend.list\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01m\u001b[Kg++:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[KparametersDefault.xxd: linker input file unused because linking not done\n",
      "\u001b[01m\u001b[Kg++:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Khtslib: linker input file unused because linking not done\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_collapseUMI_Graph.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_collapseUMIall_multiMappers.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ParametersClip_initialize.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ClipMate_clip.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ClipCR4.cpp\n",
      "cd opal && \\\n",
      "g++ -c -I./ -std=c++11   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra  -mavx2 opal.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ClipMate_clipChunk.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ClipMate_initialize.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_loadRawMatrix.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_emptyDrops_CR.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra soloInputFeatureUMI.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_countSmartSeq.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_redistributeReadsByCB.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_quantTranscript.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_sumThreads.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_countVelocyto.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_countCBgeneUMI.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome_classifyAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome_geneFullAlignOverlap_ExonOverIntron.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_cellFiltering.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_statsOutput.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra bamSortByCoordinate.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloBarcode.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ParametersSolo.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloRead.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloRead_record.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloReadBarcode.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloReadBarcode_getCBandUMI.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloBarcode_extractBarcode.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloReadFeature.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloReadFeature_record.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloReadFeature_inputRecords.cpp\n",
      "\u001b[01m\u001b[KSoloReadFeature_inputRecords.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid SoloReadFeature::inputRecords(unsigned int**, unsigned int, std::vector<unsigned int>&, std::vector<readInfoStruct>&, SoloReadFlagClass&, std::vector<unsigned int>&, std::vector<unsigned int>&)\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[KSoloReadFeature_inputRecords.cpp:21:12:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kunused variable ‘\u001b[01m\u001b[KnReadsIn\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K-Wunused-variable\u001b[m\u001b[K]\n",
      "   21 |     uint64 \u001b[01;35m\u001b[KnReadsIn\u001b[m\u001b[K = 0;\n",
      "      |            \u001b[01;35m\u001b[K^~~~~~~~\u001b[m\u001b[K\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Solo.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_outputResults.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_processRecords.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_addBAMtags.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_transformGenome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome_transformGenome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript_convertGenomeCigar.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra twoPassRunPass1.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra samHeaders.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome_genomeLoad.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome_genomeOutLoad.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript_transformGenome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_outputSpliceGraphSAM.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_mapOneReadSpliceGraph.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SpliceGraph.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SpliceGraph_swScoreSpliced.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SpliceGraph_swTraceBack.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SpliceGraph_findSuperTr.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjAlignSplit.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra GTF.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra GTF_transcriptGeneSJ.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra GTF_superTranscript.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SuperTranscriptome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_outputAlignments.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra STAR.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SharedMemory.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra PackedArray.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SuffixArrayFuns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Parameters.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Parameters_samAttributes.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra InOutStreams.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SequenceFuns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ParametersGenome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Stats.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript_alignScore.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript_generateCigarP.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Chain.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcript_variationAdjust.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Variation.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_waspMap.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_storeAligns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_stitchPieces.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_multMapSelect.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_mapOneRead.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra readLoad.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlignChunk.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlignChunk_processChunks.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlignChunk_mapChunk.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra OutSJ.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra outputSJ.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra blocksOverlap.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ThreadControl.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sysRemoveDir.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_maxMappableLength2strands.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra binarySearch2.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_outputTranscriptSAM.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_outputTranscriptSJ.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_outputTranscriptCIGARp.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_createExtendWindowsWithAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_assignAlignToWindow.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_oneRead.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_stitchWindowSeeds.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_peOverlapMergeMap.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_mappedFilter.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ParametersChimeric_initialize.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_chimericDetection.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_chimericDetectionOld.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_chimericDetectionOldOutput.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ChimericDetection.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ChimericDetection_chimericDetectionMult.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_chimericDetectionPEmerged.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra stitchWindowAligns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra extendAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra stitchAlignToTranscript.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ChimericAlign_chimericJunctionOutput.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ChimericAlign_chimericBAMoutput.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ChimericAlign_chimericStitching.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome_genomeGenerate.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra genomeParametersWrite.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra genomeScanFastaFiles.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra genomeSAindex.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Genome_insertSequences.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra insertSeqSA.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra funCompareUintAndSuffixes.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra funCompareUintAndSuffixesMemcmp.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra TimeFunctions.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ErrorWarning.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra streamFuns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra stringSubstituteAll.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome_quantAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome_geneFullAlignOverlap.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_quantTranscriptome.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Quantifications.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Transcriptome_geneCountsAddAlign.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjdbLoadFromFiles.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjdbLoadFromStream.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjdbPrepare.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjdbBuildIndex.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra sjdbInsertJunctions.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra mapThreadsSpawn.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Parameters_readFilesInit.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra Parameters_readSAMheader.cpp\n",
      "\u001b[01m\u001b[KParameters_readSAMheader.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid Parameters::readSAMheader(std::string, std::vector<std::__cxx11::basic_string<char> >)\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[KParameters_readSAMheader.cpp:32:15:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint system(const char*)\u001b[m\u001b[K’, declared with attribute warn_unused_result [\u001b[01;35m\u001b[K-Wunused-result\u001b[m\u001b[K]\n",
      "   32 |         \u001b[01;35m\u001b[Ksystem(com1.c_str())\u001b[m\u001b[K;\n",
      "      |         \u001b[01;35m\u001b[K~~~~~~^~~~~~~~~~~~~~\u001b[m\u001b[K\n",
      "g++ -c  -pipe -Wall -Wextra -O3 bam_cat.c\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra serviceFuns.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra BAMoutput.cpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra BAMfunctions.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra ReadAlign_alignBAM.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra BAMbinSortByCoordinate.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra signalFromBAM.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra bamRemoveDuplicates.cpp\n",
      "g++ -c   -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra BAMbinSortUnmapped.cpp\n",
      "g++ -o STAR  -O3 -std=c++11 -fopenmp -D'COMPILATION_TIME_PLACE=\"2022-01-17T15:30:52+01:00 :/media/Dropbox/MNM projects/snRNA-seq Synuclein project/Data analysis/scifi-analysis/STAR-2.7.10a/source\"' -D'GIT_BRANCH_COMMIT_DIFF=\"On branch main ; commit 360369fb4601a18cc34475b7434e9c042a0e0e01 ; diff files: \"' -pipe -Wall -Wextra SoloFeature_collapseUMI_Graph.o SoloFeature_collapseUMIall_multiMappers.o ParametersClip_initialize.o ClipMate_clip.o ClipCR4.o opal/opal.o ClipMate_clipChunk.o ClipMate_initialize.o SoloFeature_loadRawMatrix.o SoloFeature_emptyDrops_CR.o soloInputFeatureUMI.o SoloFeature_countSmartSeq.o SoloFeature_redistributeReadsByCB.o SoloFeature_quantTranscript.o SoloFeature_sumThreads.o SoloFeature_countVelocyto.o SoloFeature_countCBgeneUMI.o Transcriptome_classifyAlign.o Transcriptome_geneFullAlignOverlap_ExonOverIntron.o Transcriptome_alignExonOverlap.cpp SoloFeature_cellFiltering.o SoloFeature_statsOutput.o bamSortByCoordinate.o SoloBarcode.o ParametersSolo.o SoloRead.o SoloRead_record.o SoloReadBarcode.o SoloReadBarcode_getCBandUMI.o SoloBarcode_extractBarcode.o SoloReadFeature.o SoloReadFeature_record.o SoloReadFeature_inputRecords.o Solo.o SoloFeature.o SoloFeature_outputResults.o SoloFeature_processRecords.o SoloFeature_addBAMtags.o ReadAlign_transformGenome.o Genome_transformGenome.o Transcript_convertGenomeCigar.o twoPassRunPass1.o samHeaders.o Genome_genomeLoad.o Genome_genomeOutLoad.o Transcript_transformGenome.o ReadAlign_outputSpliceGraphSAM.o ReadAlign_mapOneReadSpliceGraph.o SpliceGraph.o SpliceGraph_swScoreSpliced.o SpliceGraph_swTraceBack.o SpliceGraph_findSuperTr.o sjAlignSplit.o GTF.o GTF_transcriptGeneSJ.o GTF_superTranscript.o SuperTranscriptome.o ReadAlign_outputAlignments.o ReadAlign.o STAR.o SharedMemory.o PackedArray.o SuffixArrayFuns.o Parameters.o Parameters_samAttributes.o InOutStreams.o SequenceFuns.o Genome.o ParametersGenome.o Stats.o Transcript.o Transcript_alignScore.o Transcript_generateCigarP.o Chain.o Transcript_variationAdjust.o Variation.o ReadAlign_waspMap.o ReadAlign_storeAligns.o ReadAlign_stitchPieces.o ReadAlign_multMapSelect.o ReadAlign_mapOneRead.o readLoad.o ReadAlignChunk.o ReadAlignChunk_processChunks.o ReadAlignChunk_mapChunk.o OutSJ.o outputSJ.o blocksOverlap.o ThreadControl.o sysRemoveDir.o ReadAlign_maxMappableLength2strands.o binarySearch2.o ReadAlign_outputTranscriptSAM.o ReadAlign_outputTranscriptSJ.o ReadAlign_outputTranscriptCIGARp.o ReadAlign_calcCIGAR.cpp ReadAlign_createExtendWindowsWithAlign.o ReadAlign_assignAlignToWindow.o ReadAlign_oneRead.o ReadAlign_stitchWindowSeeds.o ReadAlign_peOverlapMergeMap.o ReadAlign_mappedFilter.o ParametersChimeric_initialize.o ReadAlign_chimericDetection.o ReadAlign_chimericDetectionOld.o ReadAlign_chimericDetectionOldOutput.o ChimericDetection.o ChimericDetection_chimericDetectionMult.o ReadAlign_chimericDetectionPEmerged.o stitchWindowAligns.o extendAlign.o stitchAlignToTranscript.o ChimericSegment.cpp ChimericAlign.cpp ChimericAlign_chimericJunctionOutput.o ChimericAlign_chimericBAMoutput.o ChimericAlign_chimericStitching.o Genome_genomeGenerate.o genomeParametersWrite.o genomeScanFastaFiles.o genomeSAindex.o Genome_insertSequences.o insertSeqSA.o funCompareUintAndSuffixes.o funCompareUintAndSuffixesMemcmp.o TimeFunctions.o ErrorWarning.o streamFuns.o stringSubstituteAll.o Transcriptome.o Transcriptome_quantAlign.o Transcriptome_geneFullAlignOverlap.o ReadAlign_quantTranscriptome.o Quantifications.o Transcriptome_geneCountsAddAlign.o sjdbLoadFromFiles.o sjdbLoadFromStream.o sjdbPrepare.o sjdbBuildIndex.o sjdbInsertJunctions.o mapThreadsSpawn.o Parameters_readFilesInit.o Parameters_openReadsFiles.cpp Parameters_closeReadsFiles.cpp Parameters_readSAMheader.o bam_cat.o serviceFuns.o GlobalVariables.cpp BAMoutput.o BAMfunctions.o ReadAlign_alignBAM.o BAMbinSortByCoordinate.o signalFromBAM.o bamRemoveDuplicates.o BAMbinSortUnmapped.o  -pthread -Lhtslib -Bstatic -lhts -Bdynamic -lz \n",
      "\u001b[01m\u001b[KParameters_openReadsFiles.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid Parameters::openReadsFiles()\u001b[m\u001b[K’:\n",
      "\u001b[01m\u001b[KParameters_openReadsFiles.cpp:25:20:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[Kimate\u001b[m\u001b[K’ might be clobbered by ‘\u001b[01m\u001b[Klongjmp\u001b[m\u001b[K’ or ‘\u001b[01m\u001b[Kvfork\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K-Wclobbered\u001b[m\u001b[K]\n",
      "   25 |          for (uint \u001b[01;35m\u001b[Kimate\u001b[m\u001b[K=0;imate<readFilesNames.size();imate++) {//open readIn files\n",
      "      |                    \u001b[01;35m\u001b[K^~~~~\u001b[m\u001b[K\n"
     ]
    }
   ],
   "source": [
    "#! wget https://github.com/alexdobin/STAR/archive/2.7.10a.tar.gz\n",
    "#! tar -xzf 2.7.10a.tar.gz\n",
    "#!cd STAR-2.7.10a/source ; make STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the trimmed R3 and R1 files, save into a dict with R3 being the key and R1 being the corresponding value. \n",
    "#Iterate over all of the key-value pairs and pass them as input files in STARsolo, thus processing them 1 by one.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/media/data/AtteR/scifi-analysis/outputs_starsolo’: File exists\n",
      "\tSTAR --genomeDir /media/data/AtteR/scifi-analysis/ref_genome/rat_index --runThreadN 24 --soloType CB_UMI_Simple --soloCBwhitelist /media/data/AtteR/scifi-analysis/737K-cratac-v1.txt --soloCBstart 1 --soloCBlen 16 --soloUMIstart 17 --soloUMIlen 8 --soloBarcodeReadLength 0 --soloFeatures GeneFull --readFilesCommand zcat --soloMultiMappers Uniform EM PropUnique Unique --soloUMIdedup 1MM_All --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --genomeLoad LoadAndKeep --alignIntronMin 20 --alignIntronMax 1000000 --readFilesIn /media/data/AtteR/scifi-analysis/pilot1_output_trim/Scifi_library_2_S2_trimmed-SciFi5_oDT_R3.fastq.gz /media/data/AtteR/scifi-analysis/pilot1_output_trim/Scifi_library_2_S2_trimmed-SciFi5_oDT_R21.fastq.gz\n",
      "\tSTAR version: 2.7.9a_2021-06-25   compiled: 2021-12-21T15:48:20+01:00 :/home/lcadmin/STAR/source\n",
      "Feb 10 15:05:14 ..... started STAR run\n",
      "Feb 10 15:05:15 ..... loading genome\n"
     ]
    }
   ],
   "source": [
    "#!./runStarSolo.sh\n",
    "!./runSTARsolo_orig.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate a summary tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "star_path = \"/media/data/AtteR/scifi-analysis/pilot1_output/\"\n",
    "star_path = \"/media/data/AtteR/scifi-analysis/outputs_starsolo/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi5_oDT',\n",
       " '/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi7_oDT',\n",
       " '/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi8_oDT',\n",
       " '/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi6_oDT']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#samples_list = [dir for dir in sorted(glob.glob(star_path +\"*\")) if os.path.isdir(dir)] \n",
    "\n",
    "#samples_list = os.listdir(star_path)\n",
    "samples_list = []\n",
    "for subdir in os.listdir(star_path):\n",
    "    full_path = star_path + subdir\n",
    "    samples_list.append(full_path)\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "summaryTable = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz\n"
     ]
    },
    {
     "ename": "NotADirectoryError",
     "evalue": "[Errno 20] Not a directory: '/tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz/Log.final.out'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNotADirectoryError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-8035b14885fb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msample_id\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamples_list\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"/Log.final.out\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mntpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbasename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    678\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    679\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 680\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    681\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    682\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    573\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    574\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 575\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    577\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    931\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    932\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 933\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    934\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    935\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1215\u001b[0m             \u001b[0;31m# \"Union[str, PathLike[str], ReadCsvBuffer[bytes], ReadCsvBuffer[str]]\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1216\u001b[0m             \u001b[0;31m# , \"str\", \"bool\", \"Any\", \"Any\", \"Any\", \"Any\", \"Any\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1217\u001b[0;31m             self.handles = get_handle(  # type: ignore[call-overload]\n\u001b[0m\u001b[1;32m   1218\u001b[0m                 \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1219\u001b[0m                 \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scifi-analysis/lib/python3.8/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    787\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    788\u001b[0m             \u001b[0;31m# Encoding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 789\u001b[0;31m             handle = open(\n\u001b[0m\u001b[1;32m    790\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    791\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNotADirectoryError\u001b[0m: [Errno 20] Not a directory: '/tmp/outdata/SciFi_new_cDNA_Pool1_R21_trimPA.fastq.gz/Log.final.out'"
     ]
    }
   ],
   "source": [
    "for sample_id in samples_list :\n",
    "    print(sample_id)\n",
    "    df = pd.read_csv(sample_id + \"/Log.final.out\", sep='\\t', header=None, index_col=0)\n",
    "    print(df)\n",
    "    df = df.rename(columns={1: ntpath.basename(sample_id)})\n",
    "    summaryTable = summaryTable.merge(df,left_index=True, right_index=True, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'summaryTable' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-45ab393f113b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msummaryTable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'summaryTable' is not defined"
     ]
    }
   ],
   "source": [
    "summaryTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28\n",
      "/media/data/AtteR/scifi-analysis/summary_tables\n"
     ]
    }
   ],
   "source": [
    "from datetime import date\n",
    "run_date = date.today()\n",
    "print(run_date)\n",
    "print(output_tables)\n",
    "summaryTable.to_csv(path_or_buf=output_tables + '/ScifiPilot1_STARsoloSummary_' + str(run_date) + '.csv', sep=',', header=True, index=True, index_label='Variable')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate second summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "summaryTable = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi5_oDT\n",
      "/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi7_oDT\n",
      "/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi8_oDT\n",
      "/media/data/AtteR/scifi-analysis/outputs_starsolo/Scifi_library_2_S2_SciFi6_oDT\n"
     ]
    }
   ],
   "source": [
    "for sample_id in samples_list :\n",
    "    print(sample_id)\n",
    "    df = pd.read_csv(sample_id + \"/GeneFull/Summary.csv\", sep=',', header=None, index_col=0)\n",
    "    df = df.rename(columns={1: ntpath.basename(sample_id)})\n",
    "    summaryTable = summaryTable.merge(df,left_index=True, right_index=True, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scifi_library_2_S2_SciFi5_oDT</th>\n",
       "      <th>Scifi_library_2_S2_SciFi7_oDT</th>\n",
       "      <th>Scifi_library_2_S2_SciFi8_oDT</th>\n",
       "      <th>Scifi_library_2_S2_SciFi6_oDT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Number of Reads</th>\n",
       "      <td>1.951169e+08</td>\n",
       "      <td>1.988014e+07</td>\n",
       "      <td>1.786018e+07</td>\n",
       "      <td>4.812984e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads With Valid Barcodes</th>\n",
       "      <td>9.048160e-01</td>\n",
       "      <td>9.076040e-01</td>\n",
       "      <td>8.913700e-01</td>\n",
       "      <td>9.016370e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequencing Saturation</th>\n",
       "      <td>8.716100e-01</td>\n",
       "      <td>9.017350e-01</td>\n",
       "      <td>9.117860e-01</td>\n",
       "      <td>9.109690e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q30 Bases in CB+UMI</th>\n",
       "      <td>9.060180e-01</td>\n",
       "      <td>9.188270e-01</td>\n",
       "      <td>9.147440e-01</td>\n",
       "      <td>9.167430e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q30 Bases in RNA read</th>\n",
       "      <td>9.156260e-01</td>\n",
       "      <td>9.152880e-01</td>\n",
       "      <td>9.118650e-01</td>\n",
       "      <td>9.135630e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads Mapped to Genome: Unique+Multiple</th>\n",
       "      <td>8.232480e-01</td>\n",
       "      <td>6.736000e-01</td>\n",
       "      <td>7.762160e-01</td>\n",
       "      <td>8.040400e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads Mapped to Genome: Unique</th>\n",
       "      <td>6.065520e-01</td>\n",
       "      <td>5.318930e-01</td>\n",
       "      <td>6.259740e-01</td>\n",
       "      <td>5.669460e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads Mapped to GeneFull: Unique+Multipe GeneFull</th>\n",
       "      <td>6.499750e-01</td>\n",
       "      <td>5.111110e-01</td>\n",
       "      <td>5.801640e-01</td>\n",
       "      <td>6.235110e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reads Mapped to GeneFull: Unique GeneFull</th>\n",
       "      <td>4.775710e-01</td>\n",
       "      <td>4.034620e-01</td>\n",
       "      <td>4.670540e-01</td>\n",
       "      <td>4.290440e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Estimated Number of Cells</th>\n",
       "      <td>8.410000e+02</td>\n",
       "      <td>5.150000e+02</td>\n",
       "      <td>5.010000e+02</td>\n",
       "      <td>6.840000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unique Reads in Cells Mapped to GeneFull</th>\n",
       "      <td>8.778417e+07</td>\n",
       "      <td>6.491274e+06</td>\n",
       "      <td>7.038387e+06</td>\n",
       "      <td>1.583304e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fraction of Unique Reads in Cells</th>\n",
       "      <td>9.420710e-01</td>\n",
       "      <td>8.092970e-01</td>\n",
       "      <td>8.437630e-01</td>\n",
       "      <td>7.667400e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mean Reads per Cell</th>\n",
       "      <td>1.043800e+05</td>\n",
       "      <td>1.260400e+04</td>\n",
       "      <td>1.404800e+04</td>\n",
       "      <td>2.314700e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Median Reads per Cell</th>\n",
       "      <td>1.023450e+05</td>\n",
       "      <td>7.237000e+03</td>\n",
       "      <td>8.510000e+03</td>\n",
       "      <td>1.347300e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UMIs in Cells</th>\n",
       "      <td>1.050457e+07</td>\n",
       "      <td>4.580890e+05</td>\n",
       "      <td>4.463930e+05</td>\n",
       "      <td>1.110523e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mean UMI per Cell</th>\n",
       "      <td>1.249000e+04</td>\n",
       "      <td>8.890000e+02</td>\n",
       "      <td>8.910000e+02</td>\n",
       "      <td>1.623000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Median UMI per Cell</th>\n",
       "      <td>1.258100e+04</td>\n",
       "      <td>4.950000e+02</td>\n",
       "      <td>5.150000e+02</td>\n",
       "      <td>9.140000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mean GeneFull per Cell</th>\n",
       "      <td>3.937000e+03</td>\n",
       "      <td>5.860000e+02</td>\n",
       "      <td>5.860000e+02</td>\n",
       "      <td>9.590000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Median GeneFull per Cell</th>\n",
       "      <td>4.117000e+03</td>\n",
       "      <td>3.900000e+02</td>\n",
       "      <td>3.980000e+02</td>\n",
       "      <td>6.740000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total GeneFull Detected</th>\n",
       "      <td>2.330700e+04</td>\n",
       "      <td>1.601200e+04</td>\n",
       "      <td>1.591400e+04</td>\n",
       "      <td>1.834100e+04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Scifi_library_2_S2_SciFi5_oDT  \\\n",
       "0                                                                                  \n",
       "Number of Reads                                                     1.951169e+08   \n",
       "Reads With Valid Barcodes                                           9.048160e-01   \n",
       "Sequencing Saturation                                               8.716100e-01   \n",
       "Q30 Bases in CB+UMI                                                 9.060180e-01   \n",
       "Q30 Bases in RNA read                                               9.156260e-01   \n",
       "Reads Mapped to Genome: Unique+Multiple                             8.232480e-01   \n",
       "Reads Mapped to Genome: Unique                                      6.065520e-01   \n",
       "Reads Mapped to GeneFull: Unique+Multipe GeneFull                   6.499750e-01   \n",
       "Reads Mapped to GeneFull: Unique GeneFull                           4.775710e-01   \n",
       "Estimated Number of Cells                                           8.410000e+02   \n",
       "Unique Reads in Cells Mapped to GeneFull                            8.778417e+07   \n",
       "Fraction of Unique Reads in Cells                                   9.420710e-01   \n",
       "Mean Reads per Cell                                                 1.043800e+05   \n",
       "Median Reads per Cell                                               1.023450e+05   \n",
       "UMIs in Cells                                                       1.050457e+07   \n",
       "Mean UMI per Cell                                                   1.249000e+04   \n",
       "Median UMI per Cell                                                 1.258100e+04   \n",
       "Mean GeneFull per Cell                                              3.937000e+03   \n",
       "Median GeneFull per Cell                                            4.117000e+03   \n",
       "Total GeneFull Detected                                             2.330700e+04   \n",
       "\n",
       "                                                   Scifi_library_2_S2_SciFi7_oDT  \\\n",
       "0                                                                                  \n",
       "Number of Reads                                                     1.988014e+07   \n",
       "Reads With Valid Barcodes                                           9.076040e-01   \n",
       "Sequencing Saturation                                               9.017350e-01   \n",
       "Q30 Bases in CB+UMI                                                 9.188270e-01   \n",
       "Q30 Bases in RNA read                                               9.152880e-01   \n",
       "Reads Mapped to Genome: Unique+Multiple                             6.736000e-01   \n",
       "Reads Mapped to Genome: Unique                                      5.318930e-01   \n",
       "Reads Mapped to GeneFull: Unique+Multipe GeneFull                   5.111110e-01   \n",
       "Reads Mapped to GeneFull: Unique GeneFull                           4.034620e-01   \n",
       "Estimated Number of Cells                                           5.150000e+02   \n",
       "Unique Reads in Cells Mapped to GeneFull                            6.491274e+06   \n",
       "Fraction of Unique Reads in Cells                                   8.092970e-01   \n",
       "Mean Reads per Cell                                                 1.260400e+04   \n",
       "Median Reads per Cell                                               7.237000e+03   \n",
       "UMIs in Cells                                                       4.580890e+05   \n",
       "Mean UMI per Cell                                                   8.890000e+02   \n",
       "Median UMI per Cell                                                 4.950000e+02   \n",
       "Mean GeneFull per Cell                                              5.860000e+02   \n",
       "Median GeneFull per Cell                                            3.900000e+02   \n",
       "Total GeneFull Detected                                             1.601200e+04   \n",
       "\n",
       "                                                   Scifi_library_2_S2_SciFi8_oDT  \\\n",
       "0                                                                                  \n",
       "Number of Reads                                                     1.786018e+07   \n",
       "Reads With Valid Barcodes                                           8.913700e-01   \n",
       "Sequencing Saturation                                               9.117860e-01   \n",
       "Q30 Bases in CB+UMI                                                 9.147440e-01   \n",
       "Q30 Bases in RNA read                                               9.118650e-01   \n",
       "Reads Mapped to Genome: Unique+Multiple                             7.762160e-01   \n",
       "Reads Mapped to Genome: Unique                                      6.259740e-01   \n",
       "Reads Mapped to GeneFull: Unique+Multipe GeneFull                   5.801640e-01   \n",
       "Reads Mapped to GeneFull: Unique GeneFull                           4.670540e-01   \n",
       "Estimated Number of Cells                                           5.010000e+02   \n",
       "Unique Reads in Cells Mapped to GeneFull                            7.038387e+06   \n",
       "Fraction of Unique Reads in Cells                                   8.437630e-01   \n",
       "Mean Reads per Cell                                                 1.404800e+04   \n",
       "Median Reads per Cell                                               8.510000e+03   \n",
       "UMIs in Cells                                                       4.463930e+05   \n",
       "Mean UMI per Cell                                                   8.910000e+02   \n",
       "Median UMI per Cell                                                 5.150000e+02   \n",
       "Mean GeneFull per Cell                                              5.860000e+02   \n",
       "Median GeneFull per Cell                                            3.980000e+02   \n",
       "Total GeneFull Detected                                             1.591400e+04   \n",
       "\n",
       "                                                   Scifi_library_2_S2_SciFi6_oDT  \n",
       "0                                                                                 \n",
       "Number of Reads                                                     4.812984e+07  \n",
       "Reads With Valid Barcodes                                           9.016370e-01  \n",
       "Sequencing Saturation                                               9.109690e-01  \n",
       "Q30 Bases in CB+UMI                                                 9.167430e-01  \n",
       "Q30 Bases in RNA read                                               9.135630e-01  \n",
       "Reads Mapped to Genome: Unique+Multiple                             8.040400e-01  \n",
       "Reads Mapped to Genome: Unique                                      5.669460e-01  \n",
       "Reads Mapped to GeneFull: Unique+Multipe GeneFull                   6.235110e-01  \n",
       "Reads Mapped to GeneFull: Unique GeneFull                           4.290440e-01  \n",
       "Estimated Number of Cells                                           6.840000e+02  \n",
       "Unique Reads in Cells Mapped to GeneFull                            1.583304e+07  \n",
       "Fraction of Unique Reads in Cells                                   7.667400e-01  \n",
       "Mean Reads per Cell                                                 2.314700e+04  \n",
       "Median Reads per Cell                                               1.347300e+04  \n",
       "UMIs in Cells                                                       1.110523e+06  \n",
       "Mean UMI per Cell                                                   1.623000e+03  \n",
       "Median UMI per Cell                                                 9.140000e+02  \n",
       "Mean GeneFull per Cell                                              9.590000e+02  \n",
       "Median GeneFull per Cell                                            6.740000e+02  \n",
       "Total GeneFull Detected                                             1.834100e+04  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaryTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28\n",
      "/media/data/AtteR/scifi-analysis/summary_tables\n"
     ]
    }
   ],
   "source": [
    "from datetime import date\n",
    "run_date = date.today()\n",
    "print(run_date)\n",
    "print(output_tables)\n",
    "summaryTable.to_csv(path_or_buf=output_tables + '/ScifiPilot1_STARsoloGeneFul_' + str(run_date) + '.csv', sep=',', header=True, index=True, index_label='Variable')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze AAV sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R21_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R21_001.fastq.gz']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*AAV*R21*.fastq.gz'))\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trimBC(sample_id) :\n",
    "    print(sample_id)\n",
    "    #sample_id = trimSamples_list[10]\n",
    "    inPathR2 = base_path + ntpath.basename(sample_id).replace('_R21_001.fastq.gz','_R3_001.fastq.gz')\n",
    "    outPathR1 = export_path + ntpath.basename(sample_id).replace('_R21_001.fastq.gz','_AAV-{name}_R21.fastq.gz')\n",
    "    outPathR2 = export_path + ntpath.basename(sample_id).replace('_R21_001.fastq.gz','_AAV-{name}_R3.fastq.gz')\n",
    "    \n",
    "    trimLog = export_path + ntpath.basename(sample_id).replace('_R21_001.fastq.gz','_AAV-Log.tsv')\n",
    "    param = \"--cores=0 -g aSyn=ACTACGAACCTGAAGCCTAAAGG...CCTAATCAACCTCTGGATTACAAAATTTGTG -g tagBFP=AACTGGGGCACAAGCTTAATTAA...AGGCCTAATCAACCTCTGGATTACAAAATTT --trimmed-only --report=minimal\"\n",
    "    call_sequence = \"cutadapt \"+param+\" -o \"+outPathR2+\" -p \" + outPathR1 + \" \"+ inPathR2 + \" \" + sample_id +\" > \" + trimLog\n",
    "    !$call_sequence\n",
    "    print(call_sequence)\n",
    "    #call_sequence    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R21_001.fastq.gz\n",
      "Done           00:00:14     2,822,255 reads @   5.1 µs/read;  11.85 M reads/minute\n",
      "cutadapt --cores=0 -g aSyn=ACTACGAACCTGAAGCCTAAAGG...CCTAATCAACCTCTGGATTACAAAATTTGTG -g tagBFP=AACTGGGGCACAAGCTTAATTAA...AGGCCTAATCAACCTCTGGATTACAAAATTT --trimmed-only --report=minimal -o /tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-{name}_R3.fastq.gz -p /tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-{name}_R21.fastq.gz /tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R3_001.fastq.gz /tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R21_001.fastq.gz > /tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv\n",
      "/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R21_001.fastq.gz\n",
      "Done           00:00:08     1,651,948 reads @   5.0 µs/read;  12.06 M reads/minute\n",
      "cutadapt --cores=0 -g aSyn=ACTACGAACCTGAAGCCTAAAGG...CCTAATCAACCTCTGGATTACAAAATTTGTG -g tagBFP=AACTGGGGCACAAGCTTAATTAA...AGGCCTAATCAACCTCTGGATTACAAAATTT --trimmed-only --report=minimal -o /tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-{name}_R3.fastq.gz -p /tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-{name}_R21.fastq.gz /tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R3_001.fastq.gz /tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R21_001.fastq.gz > /tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[trimBC(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>in_reads</th>\n",
       "      <th>in_bp</th>\n",
       "      <th>too_short</th>\n",
       "      <th>too_long</th>\n",
       "      <th>too_many_n</th>\n",
       "      <th>out_reads</th>\n",
       "      <th>w/adapters</th>\n",
       "      <th>qualtrim_bp</th>\n",
       "      <th>out_bp</th>\n",
       "      <th>w/adapters2</th>\n",
       "      <th>qualtrim2_bp</th>\n",
       "      <th>out2_bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv</th>\n",
       "      <td>OK</td>\n",
       "      <td>2822255</td>\n",
       "      <td>445916290</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2648737</td>\n",
       "      <td>2648737</td>\n",
       "      <td>0</td>\n",
       "      <td>63263488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>113895691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv</th>\n",
       "      <td>OK</td>\n",
       "      <td>1651948</td>\n",
       "      <td>261007784</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1538839</td>\n",
       "      <td>1538839</td>\n",
       "      <td>0</td>\n",
       "      <td>37967841</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66170077</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     status  in_reads      in_bp  too_short  \\\n",
       "SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv       OK   2822255  445916290          0   \n",
       "SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv     OK   1651948  261007784          0   \n",
       "\n",
       "                                      too_long  too_many_n  out_reads  \\\n",
       "SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv           0           0    2648737   \n",
       "SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv         0           0    1538839   \n",
       "\n",
       "                                      w/adapters  qualtrim_bp    out_bp  \\\n",
       "SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv       2648737            0  63263488   \n",
       "SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv     1538839            0  37967841   \n",
       "\n",
       "                                      w/adapters2  qualtrim2_bp    out2_bp  \n",
       "SciFi_Kapa_AAV_aSyn_S6_AAV-Log.tsv              0             0  113895691  \n",
       "SciFi_Kapa_AAV_tagBFP_S5_AAV-Log.tsv            0             0   66170077  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_list = sorted(glob.glob(export_path+'*AAV-Log.tsv'))\n",
    "allSampleTable = pd.DataFrame()\n",
    "\n",
    "for qc_file in qc_list :\n",
    "    df=pd.read_csv(qc_file, sep='\\t') # , header=0, index_col=0,skiprows=None\n",
    "    df = df.set_index(pd.Series(ntpath.basename(qc_file).replace('_trimLog.tsv','')))\n",
    "    allSampleTable = pd.concat([allSampleTable, df], axis=0)\n",
    "\n",
    "allSampleTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn_R21.fastq.gz\n",
      "2648732\n",
      "/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn_R3.fastq.gz\n",
      "2648732\n",
      "/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-tagBFP_R21.fastq.gz\n",
      "5\n",
      "/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-tagBFP_R3.fastq.gz\n",
      "5\n",
      "/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-aSyn_R21.fastq.gz\n",
      "364\n",
      "/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-aSyn_R3.fastq.gz\n",
      "364\n",
      "/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-tagBFP_R21.fastq.gz\n",
      "1538475\n",
      "/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_AAV-tagBFP_R3.fastq.gz\n",
      "1538475\n"
     ]
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(export_path+'*AAV*.fastq.gz'))\n",
    "for sample_file in samples_list :\n",
    "    print(sample_file)\n",
    "    result = subprocess.run(['pigz -dc '+ sample_file+ ' | wc -l'], stdout=subprocess.PIPE, shell=True)\n",
    "    sampleReads = int(int(result.stdout)/4)\n",
    "    if sampleReads < 4000 :\n",
    "        os.remove(sample_file)\n",
    "    print(sampleReads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn_R3.fastq.gz']"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AAV_list = sorted(glob.glob(export_path+'*AAV-aSyn_R3.fastq.gz'))\n",
    "AAV_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runSC(sample_id) :\n",
    "    print(sample_id)\n",
    "    \n",
    "    outPath = export_path + ntpath.basename(sample_id).replace('_R3.fastq.gz','_starcode.tsv')\n",
    "    starcode_call= \"gunzip -c '\"+ sample_id +\"' | /media/data/TomasB/starcode/starcode -t 48 -o '\"+outPath + \"'\"\n",
    "    !$starcode_call "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn_R3.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[runSC(sample_id) for sample_id in AAV_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn_starcode.tsv']"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AAV_SC_list = sorted(glob.glob(export_path+'*AAV-aSyn_starcode.tsv'))\n",
    "AAV_SC_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "starcodeTable = pd.DataFrame()\n",
    "for sample_id in AAV_SC_list :\n",
    "    df = pd.read_csv(sample_id, sep='\\t', header=None, index_col=0, dtype={0: str, 1: 'Int64'})\n",
    "    df = df.rename(columns={1: ntpath.basename(sample_id).replace('_starcode.tsv','')})\n",
    "    starcodeTable = starcodeTable.merge(df,left_index=True, right_index=True, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACATCCATGCGTACATCATTCCTT</th>\n",
       "      <td>61980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAGCGCTGGAGGGAGGACAGACGT</th>\n",
       "      <td>59413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATATGCGGCAGTGTTCATTGATTC</th>\n",
       "      <td>55206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTGAATCGCGTGAACCCTTGAAT</th>\n",
       "      <td>49405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATGGACTGGTACACAGGCTGATGT</th>\n",
       "      <td>46600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTCCGCATGCAATCCTTGCTAT</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTGCGGCAGGGTTCATTTATTC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTGCTCGATGATATATACTATG</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTGGTTCCTTGTGTGCAC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTGTGGTTTGGTCGCAAGAAAT</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15487 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn\n",
       "0                                                        \n",
       "ACATCCATGCGTACATCATTCCTT                            61980\n",
       "GAGCGCTGGAGGGAGGACAGACGT                            59413\n",
       "ATATGCGGCAGTGTTCATTGATTC                            55206\n",
       "CCTGAATCGCGTGAACCCTTGAAT                            49405\n",
       "ATGGACTGGTACACAGGCTGATGT                            46600\n",
       "...                                                   ...\n",
       "TTTTCCGCATGCAATCCTTGCTAT                                1\n",
       "TTTTGCGGCAGGGTTCATTTATTC                                1\n",
       "TTTTGCTCGATGATATATACTATG                                1\n",
       "TTTTGGTTCCTTGTGTGCAC                                    1\n",
       "TTTTGTGGTTTGGTCGCAAGAAAT                                1\n",
       "\n",
       "[15487 rows x 1 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "starcodeTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "starcodeTable = starcodeTable[starcodeTable['SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn']>2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "allBCsCount = starcodeTable.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SciFi_Kapa_AAV_aSyn_S6_AAV-aSyn    296\n",
       "dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allBCsCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Kapa_input_S2_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Phusion_TAG1_N712_S3_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Phusion_TAG1_N715_S4_R1_001.fastq.gz',\n",
       " '/tmp/indata/SciFi_Phusion_input_S1_R1_001.fastq.gz']"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R1*.fastq.gz'))\n",
    "#samples_list= [item for item in samples_list if 'Undetermined' not in item]\n",
    "#samples_list= [item for item in samples_list if 'BC' in item]\n",
    "samples_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runSC(sample_id) :\n",
    "    print(sample_id)\n",
    "    #outPath = re.sub('R*[1-3]*_001.fastq.gz', 'starcode.tsv', export_path + ntpath.basename(sample_id), 2)\n",
    "    outPath = export_path + ntpath.basename(sample_id).replace('_trimPA.fastq.gz','_starcode.tsv')\n",
    "    starcode_call= \"gunzip -c '\"+ sample_id +\"' | /media/data/TomasB/starcode/starcode -t 48 -o '\"+outPath + \"'\"\n",
    "    !$starcode_call "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/outdata/SciFi_Kapa_input_S2_R1_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Kapa_input_S2_R3_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R1_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "^C\n",
      "/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R1_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Phusion_input_S1_R1_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n",
      "/tmp/outdata/SciFi_Phusion_input_S1_R3_trimPA.fastq.gz\n",
      "running starcode with 48 threads\n",
      "reading input files\n",
      "FASTQ format detected\n",
      "sorting\n",
      "setting dist to 2\n",
      "progress: 100.00%\n",
      "message passing clustering\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None, None, None, None, None]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(export_path+'*R3_trimPA.fastq.gz'))\n",
    "[runSC(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_R3_starcode.tsv',\n",
       " '/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_R3_starcode.tsv',\n",
       " '/tmp/outdata/SciFi_Kapa_input_S2_R3_starcode.tsv',\n",
       " '/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_starcode.tsv',\n",
       " '/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_starcode.tsv',\n",
       " '/tmp/outdata/SciFi_Phusion_input_S1_R3_starcode.tsv']"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "R3_list = sorted(glob.glob(export_path+'*R3_starcode.tsv'))\n",
    "R3_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "starcodeTable = pd.DataFrame()\n",
    "for sample_id in R3_list[2:] :\n",
    "    df = pd.read_csv(sample_id, sep='\\t', header=None, index_col=0, dtype={0: str, 1: 'Int64'}, nrows=15)\n",
    "    df = df.rename(columns={1: ntpath.basename(sample_id).replace('_starcode.tsv','')})\n",
    "    starcodeTable = starcodeTable.merge(df,left_index=True, right_index=True, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SciFi_Kapa_input_S2_R3</th>\n",
       "      <th>SciFi_Phusion_TAG2_N714_S3_R3</th>\n",
       "      <th>SciFi_Phusion_TAG2_N715_S4_R3</th>\n",
       "      <th>SciFi_Phusion_input_S1_R3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>4050511</td>\n",
       "      <td>1495732</td>\n",
       "      <td>1598900</td>\n",
       "      <td>53551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGC</th>\n",
       "      <td>3693705</td>\n",
       "      <td>820514</td>\n",
       "      <td>920244</td>\n",
       "      <td>28857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACC</th>\n",
       "      <td>420898</td>\n",
       "      <td>141028</td>\n",
       "      <td>153227</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCAAAAAAAAAAAAAAAAA</th>\n",
       "      <td>524815</td>\n",
       "      <td>129951</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGAAAAAAAAAAAAAAAA</th>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>167672</td>\n",
       "      <td>179724</td>\n",
       "      <td>5298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTC</th>\n",
       "      <td>14706457</td>\n",
       "      <td>4238130</td>\n",
       "      <td>4618962</td>\n",
       "      <td>133826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAAAAAAAAAAAAAA</th>\n",
       "      <td>1073009</td>\n",
       "      <td>413310</td>\n",
       "      <td>461689</td>\n",
       "      <td>11084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAG</th>\n",
       "      <td>3103904</td>\n",
       "      <td>929411</td>\n",
       "      <td>1009270</td>\n",
       "      <td>26607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGAAAAAAAAAAAA</th>\n",
       "      <td>1381805</td>\n",
       "      <td>354859</td>\n",
       "      <td>371144</td>\n",
       "      <td>13404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGACAAAAAAAAAA</th>\n",
       "      <td>582089</td>\n",
       "      <td>171124</td>\n",
       "      <td>185525</td>\n",
       "      <td>5841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGACTAAAAAAAAA</th>\n",
       "      <td>538187</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>142321</td>\n",
       "      <td>4744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGATCAAAAAAAAA</th>\n",
       "      <td>1087004</td>\n",
       "      <td>300344</td>\n",
       "      <td>326288</td>\n",
       "      <td>9886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGATT</th>\n",
       "      <td>855489</td>\n",
       "      <td>239425</td>\n",
       "      <td>259837</td>\n",
       "      <td>7545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCACCGTCAGATTAAAAAAAAA</th>\n",
       "      <td>3659567</td>\n",
       "      <td>867808</td>\n",
       "      <td>961783</td>\n",
       "      <td>33731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGCTGTCTCTTATACACATCTGACG</th>\n",
       "      <td>1497110</td>\n",
       "      <td>781631</td>\n",
       "      <td>791651</td>\n",
       "      <td>21308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGCTGTCTCTTATACACATCTGCCG</th>\n",
       "      <td>400504</td>\n",
       "      <td>204617</td>\n",
       "      <td>199349</td>\n",
       "      <td>5206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           SciFi_Kapa_input_S2_R3  \\\n",
       "0                                                   \n",
       "NaN                                       4050511   \n",
       "GCTGC                                     3693705   \n",
       "GCTGCACC                                   420898   \n",
       "GCTGCACCAAAAAAAAAAAAAAAAA                  524815   \n",
       "GCTGCACCGAAAAAAAAAAAAAAAA                    <NA>   \n",
       "GCTGCACCGTC                              14706457   \n",
       "GCTGCACCGTCAAAAAAAAAAAAAA                 1073009   \n",
       "GCTGCACCGTCAG                             3103904   \n",
       "GCTGCACCGTCAGAAAAAAAAAAAA                 1381805   \n",
       "GCTGCACCGTCAGACAAAAAAAAAA                  582089   \n",
       "GCTGCACCGTCAGACTAAAAAAAAA                  538187   \n",
       "GCTGCACCGTCAGATCAAAAAAAAA                 1087004   \n",
       "GCTGCACCGTCAGATT                           855489   \n",
       "GCTGCACCGTCAGATTAAAAAAAAA                 3659567   \n",
       "GGCTGTCTCTTATACACATCTGACG                 1497110   \n",
       "GGCTGTCTCTTATACACATCTGCCG                  400504   \n",
       "\n",
       "                           SciFi_Phusion_TAG2_N714_S3_R3  \\\n",
       "0                                                          \n",
       "NaN                                              1495732   \n",
       "GCTGC                                             820514   \n",
       "GCTGCACC                                          141028   \n",
       "GCTGCACCAAAAAAAAAAAAAAAAA                         129951   \n",
       "GCTGCACCGAAAAAAAAAAAAAAAA                         167672   \n",
       "GCTGCACCGTC                                      4238130   \n",
       "GCTGCACCGTCAAAAAAAAAAAAAA                         413310   \n",
       "GCTGCACCGTCAG                                     929411   \n",
       "GCTGCACCGTCAGAAAAAAAAAAAA                         354859   \n",
       "GCTGCACCGTCAGACAAAAAAAAAA                         171124   \n",
       "GCTGCACCGTCAGACTAAAAAAAAA                           <NA>   \n",
       "GCTGCACCGTCAGATCAAAAAAAAA                         300344   \n",
       "GCTGCACCGTCAGATT                                  239425   \n",
       "GCTGCACCGTCAGATTAAAAAAAAA                         867808   \n",
       "GGCTGTCTCTTATACACATCTGACG                         781631   \n",
       "GGCTGTCTCTTATACACATCTGCCG                         204617   \n",
       "\n",
       "                           SciFi_Phusion_TAG2_N715_S4_R3  \\\n",
       "0                                                          \n",
       "NaN                                              1598900   \n",
       "GCTGC                                             920244   \n",
       "GCTGCACC                                          153227   \n",
       "GCTGCACCAAAAAAAAAAAAAAAAA                           <NA>   \n",
       "GCTGCACCGAAAAAAAAAAAAAAAA                         179724   \n",
       "GCTGCACCGTC                                      4618962   \n",
       "GCTGCACCGTCAAAAAAAAAAAAAA                         461689   \n",
       "GCTGCACCGTCAG                                    1009270   \n",
       "GCTGCACCGTCAGAAAAAAAAAAAA                         371144   \n",
       "GCTGCACCGTCAGACAAAAAAAAAA                         185525   \n",
       "GCTGCACCGTCAGACTAAAAAAAAA                         142321   \n",
       "GCTGCACCGTCAGATCAAAAAAAAA                         326288   \n",
       "GCTGCACCGTCAGATT                                  259837   \n",
       "GCTGCACCGTCAGATTAAAAAAAAA                         961783   \n",
       "GGCTGTCTCTTATACACATCTGACG                         791651   \n",
       "GGCTGTCTCTTATACACATCTGCCG                         199349   \n",
       "\n",
       "                           SciFi_Phusion_input_S1_R3  \n",
       "0                                                     \n",
       "NaN                                            53551  \n",
       "GCTGC                                          28857  \n",
       "GCTGCACC                                        <NA>  \n",
       "GCTGCACCAAAAAAAAAAAAAAAAA                       4391  \n",
       "GCTGCACCGAAAAAAAAAAAAAAAA                       5298  \n",
       "GCTGCACCGTC                                   133826  \n",
       "GCTGCACCGTCAAAAAAAAAAAAAA                      11084  \n",
       "GCTGCACCGTCAG                                  26607  \n",
       "GCTGCACCGTCAGAAAAAAAAAAAA                      13404  \n",
       "GCTGCACCGTCAGACAAAAAAAAAA                       5841  \n",
       "GCTGCACCGTCAGACTAAAAAAAAA                       4744  \n",
       "GCTGCACCGTCAGATCAAAAAAAAA                       9886  \n",
       "GCTGCACCGTCAGATT                                7545  \n",
       "GCTGCACCGTCAGATTAAAAAAAAA                      33731  \n",
       "GGCTGTCTCTTATACACATCTGACG                      21308  \n",
       "GGCTGTCTCTTATACACATCTGCCG                       5206  "
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "starcodeTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "starcodeTable.to_csv(path_or_buf=export_path+ 'R3_starcodeTable.tsv', sep='\\t', header=True, index=True, index_label='Sequence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runCollapser(sample_id) :\n",
    "    print(sample_id)\n",
    "    outPath = export_path + ntpath.basename(sample_id).replace('_001.fastq.gz','_Collapsed.fastq.gz')\n",
    "    outClusterPath = export_path + ntpath.basename(sample_id).replace('_001.fastq.gz','_Clusters.txt')\n",
    "    collapser_call= \"dedupe.sh in='\"+ sample_id +\"' out='\"+outPath +\"' csf='\" + outClusterPath + \"'\"\n",
    "    !$collapser_call "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R3_001.fastq.gz\n",
      "java -ea -Xmx204824m -Xms204824m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R3_001.fastq.gz out=/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Kapa_AAV_aSyn_S6_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Kapa_AAV_aSyn_S6_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=214781m, total=214781m, free=214731m, used=50m\n",
      "\n",
      "Found 1154137 duplicates.\n",
      "Finished exact matches.    Time: 4.362 seconds.\n",
      "Memory: max=214781m, total=214781m, free=211409m, used=3372m\n",
      "\n",
      "Found 0 contained sequences.\n",
      "Finished containment.      Time: 5.426 seconds.\n",
      "Memory: max=214781m, total=214781m, free=210554m, used=4227m\n",
      "\n",
      "Removed 0 invalid entries.\n",
      "Finished invalid removal.  Time: 0.092 seconds.\n",
      "Memory: max=214781m, total=214781m, free=210554m, used=4227m\n",
      "\n",
      "Input:                  \t2270099 reads \t\t261061385 bases.\n",
      "Duplicates:             \t1154137 reads (50.84%) \t132725755 bases (50.84%)     \t0 collisions.\n",
      "Containments:           \t0 reads (0.00%) \t0 bases (0.00%)    \t1235498134 collisions.\n",
      "Result:                 \t1115962 reads (49.16%) \t128335630 bases (49.16%)\n",
      "\n",
      "Printed output.            Time: 2.264 seconds.\n",
      "Memory: max=214781m, total=214781m, free=210050m, used=4731m\n",
      "\n",
      "Time:                         \t12.174 seconds.\n",
      "Reads Processed:       2270k \t186.47k reads/sec\n",
      "Bases Processed:        261m \t21.44m bases/sec\n",
      "/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R3_001.fastq.gz\n",
      "java -ea -Xmx204821m -Xms204821m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R3_001.fastq.gz out=/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Kapa_AAV_tagBFP_S5_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Kapa_AAV_tagBFP_S5_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=214781m, total=214781m, free=214748m, used=33m\n",
      "\n",
      "Found 679436 duplicates.\n",
      "Finished exact matches.    Time: 1.958 seconds.\n",
      "Memory: max=214781m, total=214781m, free=213070m, used=1711m\n",
      "\n",
      "Found 0 contained sequences.\n",
      "Finished containment.      Time: 1.552 seconds.\n",
      "Memory: max=214781m, total=214781m, free=212231m, used=2550m\n",
      "\n",
      "Removed 0 invalid entries.\n",
      "Finished invalid removal.  Time: 0.050 seconds.\n",
      "Memory: max=214781m, total=214781m, free=212231m, used=2550m\n",
      "\n",
      "Input:                  \t966321 reads \t\t111126915 bases.\n",
      "Duplicates:             \t679436 reads (70.31%) \t78135140 bases (70.31%)     \t0 collisions.\n",
      "Containments:           \t0 reads (0.00%) \t0 bases (0.00%)    \t240403724 collisions.\n",
      "Result:                 \t286885 reads (29.69%) \t32991775 bases (29.69%)\n",
      "\n",
      "Printed output.            Time: 0.752 seconds.\n",
      "Memory: max=214781m, total=214781m, free=212064m, used=2717m\n",
      "\n",
      "Time:                         \t4.341 seconds.\n",
      "Reads Processed:        966k \t222.61k reads/sec\n",
      "Bases Processed:        111m \t25.60m bases/sec\n",
      "/tmp/indata/SciFi_Kapa_input_S2_R3_001.fastq.gz\n",
      "java -ea -Xmx204822m -Xms204822m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Kapa_input_S2_R3_001.fastq.gz out=/tmp/outdata/SciFi_Kapa_input_S2_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Kapa_input_S2_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Kapa_input_S2_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Kapa_input_S2_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Kapa_input_S2_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=214781m, total=214781m, free=214748m, used=33m\n",
      "\n",
      "Found 1651471 duplicates.\n",
      "Finished exact matches.    Time: 109.148 seconds.\n",
      "Memory: max=214781m, total=214781m, free=175348m, used=39433m\n",
      "\n",
      "^C\n",
      "/tmp/indata/SciFi_Phusion_TAG2_N714_S3_R3_001.fastq.gz\n",
      "java -ea -Xmx168648m -Xms168648m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Phusion_TAG2_N714_S3_R3_001.fastq.gz out=/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Phusion_TAG2_N714_S3_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Phusion_TAG2_N714_S3_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=176865m, total=176865m, free=176815m, used=50m\n",
      "\n",
      "^C\n",
      "/tmp/indata/SciFi_Phusion_TAG2_N715_S4_R3_001.fastq.gz\n",
      "java -ea -Xmx203374m -Xms203374m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Phusion_TAG2_N715_S4_R3_001.fastq.gz out=/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Phusion_TAG2_N715_S4_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Phusion_TAG2_N715_S4_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=213271m, total=213271m, free=213238m, used=33m\n",
      "\n",
      "^C\n",
      "/tmp/indata/SciFi_Phusion_input_S1_R3_001.fastq.gz\n",
      "java -ea -Xmx202712m -Xms202712m -cp /usr/local/bin/bbmap/current/ jgi.Dedupe in=/tmp/indata/SciFi_Phusion_input_S1_R3_001.fastq.gz out=/tmp/outdata/SciFi_Phusion_input_S1_R3_Collapsed.fastq.gz csf=/tmp/outdata/SciFi_Phusion_input_S1_R3_Clusters.txt\n",
      "Executing jgi.Dedupe [in=/tmp/indata/SciFi_Phusion_input_S1_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Phusion_input_S1_R3_Collapsed.fastq.gz, csf=/tmp/outdata/SciFi_Phusion_input_S1_R3_Clusters.txt]\n",
      "Version 38.93\n",
      "\n",
      "Initial:\n",
      "Memory: max=212567m, total=212567m, free=212516m, used=51m\n",
      "\n",
      "Found 9480 duplicates.\n",
      "Finished exact matches.    Time: 1.153 seconds.\n",
      "Memory: max=212567m, total=212567m, free=211158m, used=1409m\n",
      "\n",
      "Found 0 contained sequences.\n",
      "Finished containment.      Time: 18.229 seconds.\n",
      "Memory: max=212567m, total=212567m, free=210319m, used=2248m\n",
      "\n",
      "Removed 0 invalid entries.\n",
      "Finished invalid removal.  Time: 0.043 seconds.\n",
      "Memory: max=212567m, total=212567m, free=210319m, used=2248m\n",
      "\n",
      "Input:                  \t458051 reads \t\t52675865 bases.\n",
      "Duplicates:             \t9480 reads (2.07%) \t1090200 bases (2.07%)     \t0 collisions.\n",
      "Containments:           \t0 reads (0.00%) \t0 bases (0.00%)    \t3263482440 collisions.\n",
      "Result:                 \t448571 reads (97.93%) \t51585665 bases (97.93%)\n",
      "\n",
      "Printed output.            Time: 0.932 seconds.\n",
      "Memory: max=212567m, total=212567m, free=210084m, used=2483m\n",
      "\n",
      "Time:                         \t20.385 seconds.\n",
      "Reads Processed:        458k \t22.47k reads/sec\n",
      "Bases Processed:      52675k \t2.58m bases/sec\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None, None, None]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R3*.fastq.gz'))\n",
    "[runCollapser(sample_id) for sample_id in samples_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runBBmerge(sample_id) :\n",
    "    print(sample_id)\n",
    "    unmatched = export_path + ntpath.basename(sample_id).replace('_001.fastq.gz','_unMatched.fastq.gz')\n",
    "    samplePair = base_path + ntpath.basename(sample_id).replace('R1_001.fastq.gz','R3_001.fastq.gz')\n",
    "    unmatchedPair= export_path + ntpath.basename(samplePair).replace('_001.fastq.gz','_unMatched.fastq.gz')\n",
    "    outPath = export_path + ntpath.basename(sample_id).replace('_001.fastq.gz','_MergedPair.fastq.gz')\n",
    "    \n",
    "    outhistogram = export_path + ntpath.basename(sample_id).replace('_001.fastq.gz','_histogram.txt')\n",
    "    merge_call= \"bbmerge.sh in1='\"+ sample_id +\"' in2='\"+samplePair +\"' out='\"+outPath +\"' outu1='\"+unmatched +\"' outu2='\"+unmatchedPair +\"' ihist='\" + outhistogram + \"' adapter1=AATGATACGGCGACCACCGAGATCTACAC adapter2=CAAGCAGAAGACGGCATACGAGAT\"\n",
    "    !$merge_call "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/indata/SciFi_Kapa_input_S2_R1_001.fastq.gz\n",
      "java -ea -Xmx1000m -Xms1000m -Djava.library.path=/usr/local/bin/bbmap/jni/ -cp /usr/local/bin/bbmap/current/ jgi.BBMerge in1=/tmp/indata/SciFi_Kapa_input_S2_R1_001.fastq.gz in2=/tmp/indata/SciFi_Kapa_input_S2_R3_001.fastq.gz out=/tmp/outdata/SciFi_Kapa_input_S2_R1_MergedPair.fastq.gz outu1=/tmp/outdata/SciFi_Kapa_input_S2_R1_unMatched.fastq.gz outu2=/tmp/outdata/SciFi_Kapa_input_S2_R3_unMatched.fastq.gz ihist=/tmp/outdata/SciFi_Kapa_input_S2_R1_histogram.txt adapter1=AATGATACGGCGACCACCGAGATCTACAC adapter2=CAAGCAGAAGACGGCATACGAGAT\n",
      "Executing jgi.BBMerge [in1=/tmp/indata/SciFi_Kapa_input_S2_R1_001.fastq.gz, in2=/tmp/indata/SciFi_Kapa_input_S2_R3_001.fastq.gz, out=/tmp/outdata/SciFi_Kapa_input_S2_R1_MergedPair.fastq.gz, outu1=/tmp/outdata/SciFi_Kapa_input_S2_R1_unMatched.fastq.gz, outu2=/tmp/outdata/SciFi_Kapa_input_S2_R3_unMatched.fastq.gz, ihist=/tmp/outdata/SciFi_Kapa_input_S2_R1_histogram.txt, adapter1=AATGATACGGCGACCACCGAGATCTACAC, adapter2=CAAGCAGAAGACGGCATACGAGAT]\n",
      "Version 38.93\n",
      "\n",
      "Writing mergable reads merged.\n",
      "Started output threads.\n",
      "Total time: 90.584 seconds.\n",
      "\n",
      "Pairs:               \t43679712\n",
      "Joined:              \t2197       \t0.005%\n",
      "Ambiguous:           \t43250399   \t99.017%\n",
      "No Solution:         \t427116   \t0.978%\n",
      "Too Short:           \t0       \t0.000%\n",
      "Adapters Expected:   \t427202   \t0.489%\n",
      "Adapters Found:      \t86       \t0.000%\n",
      "\n",
      "Avg Insert:          \t120.6\n",
      "Standard Deviation:  \t5.8\n",
      "Mode:                \t114\n",
      "\n",
      "Insert range:        \t105 - 132\n",
      "90th percentile:     \t129\n",
      "75th percentile:     \t125\n",
      "50th percentile:     \t120\n",
      "25th percentile:     \t116\n",
      "10th percentile:     \t113\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_list = sorted(glob.glob(base_path+'*R1*.fastq.gz'))\n",
    "[runBBmerge(sample_id) for sample_id in samples_list[2:3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bbmerge-auto.sh in=reads.fq out=merged.fq adapter1=AATGATACGGCGACCACCGAGATCTACAC adapter2=CAAGCAGAAGACGGCATACGAGAT rem k=62 extend2=50 ecct\n",
    "\n",
    "\n",
    "\n",
    "bbmerge.sh  in1= out= outu1=unmerged.fq ihist=ihist.txt\n",
    "\n",
    "\n",
    "reformat.sh\n",
    "\n",
    "fuse.sh in1=read1.fq in2=read2.fq out=fused.fq fusepairs pad=14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'AAV_SC_list' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_41488/3777484728.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mR1_list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mstarcodeTable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0msample_id\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mAAV_SC_list\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'Int64'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m#Analyze just the 10 most common\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'AAV_SC_list' is not defined"
     ]
    }
   ],
   "source": [
    "R1_list = sorted(glob.glob(export_path+'*R1_starcode.tsv'))\n",
    "R1_list\n",
    "\n",
    "starcodeTable = pd.DataFrame()\n",
    "for sample_id in R1_list :\n",
    "    df = pd.read_csv(sample_id, sep='\\t', header=None, index_col=0, dtype={0: str, 1: 'Int64'})\n",
    "    df = df[0:10] #Analyze just the 10 most common\n",
    "    df = df.rename(columns={1: ntpath.basename(sample_id).replace('_starcode.tsv','')})\n",
    "    starcodeTable = starcodeTable.merge(df,left_index=True, right_index=True, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "starcodeTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR \n",
    "--genomeDir /home/genomes/rat/STARidx77 \n",
    "--runThreadN 24 \n",
    "--soloType CB_UMI_Simple \n",
    "--soloCBwhitelist /home/data/soloWhitelist.txt \n",
    "--soloCBstart 9 \n",
    "--soloCBlen 13 \n",
    "--soloUMIstart 1 \n",
    "--soloUMIlen 8 \n",
    "--soloBarcodeReadLength 0 \n",
    "--soloFeatures GeneFull \n",
    "--readFilesCommand zcat \n",
    "--soloMultiMappers PropUnique --outSAMtype BAM SortedByCoordinate --outSAMattributes UR CB UB GX GN --limitBAMsortRAM 220000000000 --soloCellFilter TopCells 1 0.99 10 --soloUMIdedup 1MM_All --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --genomeLoad LoadAndKeep --soloOutFileNames Scifi_library_SciFi5/ features.tsv barcodes.tsv matrix.mtx --alignIntronMin 20 --alignIntronMax 1000000 --readFilesIn /home/data/output_cutadapt_novaseq/Scifi_library_trimmed-SciFi5_oDT_R3.fastq.gz /home/data/output_cutadapt_novaseq/Scifi_library_trimmed-SciFi5_oDT_R1.fastq.gz"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
